// @myriaddreamin/typst.react@0.6.1-rc5/dist/typst.react.mjs
var ur = Object.defineProperty;
var dr = (r, e, t) => e in r ? ur(r, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : r[e] = t;
var b = (r, e, t) => dr(r, typeof e != "symbol" ? e + "" : e, t);
import { jsxs as gr, jsx as et } from "react/jsx-runtime";
import { useRef as lr, useEffect as fr } from "react";
let Ue, Le, Ot = !1;
function Lt() {
  return Ot ? Ue : void 0;
}
function Tt(r, e) {
  const t = Ue || r();
  return Le !== void 0 ? Le : Le = (async () => (Ot = !0, await t.init(e), Ue = t))();
}
function Mt(r, e, t, n) {
  const i = Lt();
  if (i) {
    t(i);
    return;
  }
  Tt(r, e).then(t).catch(n);
}
const wr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  createGlobalRenderer: Tt,
  getGlobalRenderer: Lt,
  withGlobalRenderer: Mt
}, Symbol.toStringTag, { value: "Module" })), pr = [
  "DejaVuSansMono-Bold.ttf",
  "DejaVuSansMono-BoldOblique.ttf",
  "DejaVuSansMono-Oblique.ttf",
  "DejaVuSansMono.ttf",
  "LibertinusSerif-Bold.otf",
  "LibertinusSerif-BoldItalic.otf",
  "LibertinusSerif-Italic.otf",
  "LibertinusSerif-Regular.otf",
  "LibertinusSerif-Semibold.otf",
  "LibertinusSerif-SemiboldItalic.otf",
  "NewCM10-Bold.otf",
  "NewCM10-BoldItalic.otf",
  "NewCM10-Italic.otf",
  "NewCM10-Regular.otf",
  "NewCMMath-Bold.otf",
  "NewCMMath-Book.otf",
  "NewCMMath-Regular.otf"
], br = [
  "InriaSerif-Bold.ttf",
  "InriaSerif-BoldItalic.ttf",
  "InriaSerif-Italic.ttf",
  "InriaSerif-Regular.ttf",
  "Roboto-Regular.ttf",
  "NotoSerifCJKsc-Regular.otf"
], hr = ["TwitterColorEmoji.ttf", "NotoColorEmoji-Regular-COLR.subset.ttf"];
function mr() {
  return Oe([], { assets: !1 });
}
function yr(r) {
  return Oe([], r);
}
function vr(r) {
  var t, n;
  const e = [];
  if (r && (r == null ? void 0 : r.assets) !== !1 && ((t = r == null ? void 0 : r.assets) != null && t.length) && ((n = r == null ? void 0 : r.assets) == null ? void 0 : n.length) > 0) {
    let i = {
      text: "https://cdn.jsdelivr.net/gh/typst/typst-assets@v0.13.1/files/fonts/",
      _: "https://cdn.jsdelivr.net/gh/typst/typst-dev-assets@v0.13.1/files/fonts/"
    }, s = r.assetUrlPrefix ?? i;
    typeof s == "string" ? s = { _: s } : s = { ...i, ...s };
    for (const a of Object.keys(s)) {
      const d = s[a];
      d[d.length - 1] !== "/" && (s[a] = d + "/");
    }
    const o = (a, d) => d.map((g) => (s[a] || s._) + g);
    for (const a of r.assets)
      switch (a) {
        case "text":
          e.push(...o(a, pr));
          break;
        case "cjk":
          e.push(...o(a, br));
          break;
        case "emoji":
          e.push(...o(a, hr));
          break;
      }
  }
  return e;
}
function Oe(r, e) {
  const t = vr(e), n = async (i, { ref: s, builder: o }) => {
    e != null && e.fetcher && s.setFetcher(e.fetcher), await s.loadFonts(o, [...r, ...t]);
  };
  return n._preloadRemoteFontOptions = e, n._kind = "fontLoader", n;
}
function Te(r) {
  return async (e, { builder: t }) => new Promise((n) => {
    t.set_package_registry(r, function(i) {
      return r.resolve(i, this);
    }), n();
  });
}
function tt(r) {
  return async (e, t) => {
    var n;
    if (t.alreadySetAccessModel)
      throw new Error(`already set some assess model before: ${(n = t.alreadySetAccessModel.constructor) == null ? void 0 : n.name}(${t.alreadySetAccessModel})`);
    return t.alreadySetAccessModel = r, new Promise((i) => {
      t.builder.set_access_model(r, (s) => {
        const o = r.getMTime(s);
        return o ? o.getTime() : 0;
      }, (s) => r.isFile(s) || !1, (s) => r.getRealPath(s) || s, (s) => r.readAll(s)), i();
    });
  };
}
const S = Symbol.for("reflexo-obj");
var Re;
(function(r) {
  r[r.PIXEL_PER_PT = 3] = "PIXEL_PER_PT";
})(Re || (Re = {}));
class kr {
  constructor(e, t, n) {
    b(this, "pageInfos");
    b(this, "loadPageCount");
    b(this, "imageScaleFactor");
    b(this, "container");
    b(this, "canvasList");
    b(this, "textLayerList");
    b(this, "commonList");
    b(this, "textLayerParentList");
    b(this, "semanticLayerList");
    this.pageInfos = e, this.imageScaleFactor = n.pixelPerPt ?? Re.PIXEL_PER_PT, t.innerHTML = "", t.style.width = "100%", this.container = t, this.canvasList = new Array(this.loadPageCount), this.textLayerList = new Array(this.loadPageCount), this.commonList = new Array(this.loadPageCount), this.textLayerParentList = new Array(this.loadPageCount), this.semanticLayerList = new Array(this.loadPageCount);
    const i = (s, o, a) => {
      const d = Math.ceil(o.width) * this.imageScaleFactor, g = Math.ceil(o.height) * this.imageScaleFactor, f = this.canvasList[s] = document.createElement("canvas"), m = this.semanticLayerList[s] = document.createElement("div"), C = this.textLayerList[s] = document.createElement("div"), E = this.textLayerParentList[s] = document.createElement("div");
      if (f.getContext("2d")) {
        const $ = document.createElement("div");
        f.width = d, f.height = g, $.appendChild(f), a.appendChild($), $.style.position = "absolute";
      }
      {
        E.appendChild(C), E.className = "typst-html-semantics";
        const $ = t.offsetWidth, z = $ / o.width;
        E.style.width = `${$}px`, E.style.height = `${o.height * z}px`, E.style.setProperty("--data-text-width", `${z}px`), E.style.setProperty("--data-text-height", `${z}px`), a.classList.add("typst-page"), a.classList.add("canvas"), a.style.width = `${$}px`, a.style.height = `${g * z}px`, a.style.position = "relative", m.appendChild(E), a.appendChild(m);
      }
    };
    for (let s = 0; s < this.pageInfos.length; s++) {
      const o = this.pageInfos[s];
      let a;
      a = this.commonList[s] = document.createElement("div"), t.appendChild(a), i(s, o, a);
    }
  }
  resetLayout() {
    for (let e = 0; e < this.pageInfos.length; e++) {
      const t = this.pageInfos[e], n = Math.ceil(t.width) * this.imageScaleFactor, i = Math.ceil(t.height) * this.imageScaleFactor, s = this.canvasList[e].parentElement;
      if (!s)
        throw new Error(`canvasDiv is null for page ${e}, canvas list length ${this.canvasList.length}`);
      const o = this.commonList[e], a = this.textLayerParentList[e], d = this.container.offsetWidth, g = d / n;
      a.style.width = `${d}px`, a.style.height = `${i * g}px`, o.style.width = `${d}px`, o.style.height = `${i * g}px`;
      const f = this.container.offsetWidth / n;
      s.style.transformOrigin = "0px 0px", s.style.transform = `scale(${f})`;
    }
  }
}
const xr = (r) => {
  let e = !1, t;
  return () => e ? t : (e = !0, t = r());
};
class Dt {
  constructor(e) {
    b(this, "wasmBin");
    b(this, "initOnce");
    if (typeof e != "function")
      throw new Error("initFn is not a function");
    this.initOnce = xr(async () => {
      await e(this.wasmBin);
    });
  }
  async init(e) {
    this.wasmBin = e, await this.initOnce();
  }
}
class Sr {
  constructor() {
    b(this, "loadedFonts", /* @__PURE__ */ new Set());
    b(this, "fetcher", fetch);
  }
  setFetcher(e) {
    this.fetcher = e;
  }
  async loadFonts(e, t) {
    const n = new Function("m", "return import(m)"), i = this.fetcher || (this.fetcher = await async function() {
      const { fetchBuilder: a, FileSystemCache: d } = await n("node-fetch-cache"), g = new d({
        /// By default, we don't have a complicated cache policy.
        cacheDirectory: ".cache/typst/fonts"
      }), f = a.withCache(g);
      return function(m, C) {
        const E = setTimeout(() => {
          console.warn("font fetching is stucking:", m);
        }, 15e3);
        return f(m, C).finally(() => {
          clearTimeout(E);
        });
      };
    }()), s = t.filter((a) => a instanceof Uint8Array || typeof a == "object" && "info" in a ? !0 : this.loadedFonts.has(a) ? !1 : (this.loadedFonts.add(a), !0)), o = await Promise.all(s.map(async (a) => {
      if (a instanceof Uint8Array) {
        await e.add_raw_font(a);
        return;
      }
      if (typeof a == "object" && "info" in a) {
        await e.add_lazy_font(a, "blob" in a ? a.blob : jt(a));
        return;
      }
      return new Uint8Array(await (await i(a)).arrayBuffer());
    }));
    for (const a of o)
      a && await e.add_raw_font(a);
  }
  async build(e, t, n) {
    const i = { ref: this, builder: t, hooks: n };
    for (const s of (e == null ? void 0 : e.beforeBuild) ?? [])
      await s(void 0, i);
    return n.latelyBuild && n.latelyBuild(i), await t.build();
  }
}
async function $t(r, e, t, n) {
  var i;
  return await e.init((i = r == null ? void 0 : r.getModule) == null ? void 0 : i.call(r)), await new Sr().build(r, new t(), n);
}
function jt(r) {
  return () => {
    const e = new XMLHttpRequest();
    return e.overrideMimeType("text/plain; charset=x-user-defined"), e.open("GET", r.url, !1), e.send(null), e.status === 200 && (e.response instanceof String || typeof e.response == "string") ? Uint8Array.from(e.response, (t) => t.charCodeAt(0)) : new Uint8Array();
  };
}
var G;
(function(r) {
  r[r.Doc = 0] = "Doc", r[r.Slide = 1] = "Slide";
})(G || (G = {}));
class Ze {
  constructor(e) {
    b(this, "hookedElem");
    b(this, "kModule");
    b(this, "opts");
    b(this, "modes", []);
    /// Configuration fields
    /// enable partial rendering
    b(this, "partialRendering", !0);
    /// underlying renderer
    b(this, "renderMode", "svg");
    b(this, "r");
    /// preview mode
    b(this, "previewMode", G.Doc);
    /// whether this is a content preview
    b(this, "isContentPreview", !1);
    /// whether this content preview will mix outline titles
    b(this, "isMixinOutline", !1);
    /// background color
    b(this, "backgroundColor", "black");
    /// default page color (empty string means transparent)
    b(this, "pageColor", "white");
    /// pixel per pt
    b(this, "pixelPerPt", 3);
    /// customized way to retrieving dom state
    b(this, "retrieveDOMState");
    /// State fields
    /// whether svg is updating (in triggerSvgUpdate)
    b(this, "isRendering", !1);
    /// whether kModule is initialized
    b(this, "moduleInitialized", !1);
    /// patch queue for updating data.
    b(this, "patchQueue", []);
    /// resources to dispose
    b(this, "disposeList", []);
    /// canvas render ctoken
    b(this, "canvasRenderCToken");
    /// There are two scales in this class: The real scale is to adjust the size
    /// of `hookedElem` to fit the svg. The virtual scale (scale ratio) is to let
    /// user zoom in/out the svg. For example:
    /// + the default value of virtual scale is 1, which means the svg is totally
    ///   fit in `hookedElem`.
    /// + if user set virtual scale to 0.5, then the svg will be zoomed out to fit
    ///   in half width of `hookedElem`. "real" current scale of `hookedElem`
    b(this, "currentRealScale", 1);
    /// "virtual" current scale of `hookedElem`
    b(this, "currentScaleRatio", 1);
    /// timeout for delayed viewport change
    b(this, "vpTimeout");
    /// sampled by last render time.
    b(this, "sampledRenderTime", 0);
    /// page to partial render
    b(this, "partialRenderPage", 0);
    /// outline data
    b(this, "outline");
    /// cursor position in form of [page, x, y]
    b(this, "cursorPosition");
    // id: number = rnd++;
    /// Cache fields
    /// cached state of container, default to retrieve state from `this.hookedElem`
    b(this, "cachedDOMState", {
      width: 0,
      height: 0,
      window: {
        innerWidth: 0,
        innerHeight: 0
      },
      boundingRect: {
        left: 0,
        top: 0,
        right: 0
      }
    });
    var t, n;
    this.hookedElem = e.hookedElem, this.kModule = e.kModule, this.opts = e || {};
    {
      const { renderMode: i, previewMode: s, isContentPreview: o, retrieveDOMState: a } = e || {};
      this.partialRendering = !1, this.renderMode = i ?? this.renderMode, this.previewMode = s ?? this.previewMode, this.isContentPreview = o || !1, this.retrieveDOMState = a ?? (() => ({
        width: this.hookedElem.offsetWidth,
        height: this.hookedElem.offsetHeight,
        window: {
          innerWidth: window.innerWidth,
          innerHeight: window.innerHeight
        },
        boundingRect: this.hookedElem.getBoundingClientRect()
      })), this.backgroundColor = getComputedStyle(document.documentElement).getPropertyValue("--typst-preview-background-color");
    }
    this.hookedElem.classList.add("hide-scrollbar-x"), (t = this.hookedElem.parentElement) == null || t.classList.add("hide-scrollbar-x"), this.previewMode === G.Slide && (this.hookedElem.classList.add("hide-scrollbar-y"), (n = this.hookedElem.parentElement) == null || n.classList.add("hide-scrollbar-y")), this.installCtrlWheelHandler();
  }
  reset() {
    this.kModule.reset(), this.moduleInitialized = !1;
  }
  dispose() {
    const e = this.disposeList;
    this.disposeList = [], e.forEach((t) => t());
  }
  static derive(e, t) {
    return ["rescale", "rerender", "postRender"].reduce((n, i) => (n[i] = e[`${i}$${t}`].bind(e), console.assert(n[i] !== void 0, `${i}$${t} is undefined`), n), {});
  }
  registerMode(e) {
    const t = Ze.derive(this, e);
    this.modes.push([e, t]), e === this.renderMode && (this.r = t);
  }
  installCtrlWheelHandler() {
    const e = [
      0.1,
      0.2,
      0.3,
      0.4,
      0.5,
      0.6,
      0.7,
      0.8,
      0.9,
      1,
      1.1,
      1.3,
      1.5,
      1.7,
      1.9,
      2.1,
      2.4,
      2.7,
      3,
      3.3,
      3.7,
      4.1,
      4.6,
      5.1,
      5.7,
      6.3,
      7,
      7.7,
      8.5,
      9.4,
      10
    ], t = (n) => {
      var i, s, o, a;
      if (n.ctrlKey) {
        n.preventDefault(), this.cachedDOMState = this.retrieveDOMState(), window.onresize !== null && (window.onresize = null);
        const d = this.currentScaleRatio;
        if (n.deltaY < 0) {
          if (this.currentScaleRatio >= e.at(-1))
            return;
          this.currentScaleRatio = e.filter((E) => E > this.currentScaleRatio).at(0);
        } else if (n.deltaY > 0) {
          if (this.currentScaleRatio <= e.at(0))
            return;
          this.currentScaleRatio = e.filter((E) => E < this.currentScaleRatio).at(-1);
        } else
          return;
        const g = this.currentScaleRatio / d, f = n.pageX * (g - 1), m = n.pageY * (g - 1);
        Math.abs(this.currentScaleRatio - 1) < 1e-5 ? (this.hookedElem.classList.add("hide-scrollbar-x"), (i = this.hookedElem.parentElement) == null || i.classList.add("hide-scrollbar-x"), this.previewMode === G.Slide && (this.hookedElem.classList.add("hide-scrollbar-y"), (s = this.hookedElem.parentElement) == null || s.classList.add("hide-scrollbar-y"))) : (this.hookedElem.classList.remove("hide-scrollbar-x"), (o = this.hookedElem.parentElement) == null || o.classList.remove("hide-scrollbar-x"), this.previewMode === G.Slide && (this.hookedElem.classList.remove("hide-scrollbar-y"), (a = this.hookedElem.parentElement) == null || a.classList.remove("hide-scrollbar-y")));
        const C = this.hookedElem.firstElementChild;
        if (C) {
          const E = this.getSvgScaleRatio(), D = Number.parseFloat(C.getAttribute("data-height")), $ = Math.ceil(D * E);
          this.hookedElem.style.height = `${$ * 2}px`;
        }
        return window.scrollBy(f, m), this.addViewportChange(), !1;
      }
    };
    this.renderMode !== "dom" && (typeof acquireVsCodeApi < "u" ? (window.addEventListener("wheel", t, {
      passive: !1
    }), this.disposeList.push(() => {
      window.removeEventListener("wheel", t);
    })) : (document.body.addEventListener("wheel", t, {
      passive: !1
    }), this.disposeList.push(() => {
      document.body.removeEventListener("wheel", t);
    })));
  }
  /// Get current scale from html to svg
  // Note: one should retrieve dom state before rescale
  getSvgScaleRatio() {
    const e = this.hookedElem.firstElementChild;
    if (!e)
      return 0;
    const t = this.cachedDOMState, n = Number.parseFloat(e.getAttribute("data-width") || e.getAttribute("width") || "1"), i = Number.parseFloat(e.getAttribute("data-height") || e.getAttribute("height") || "1");
    return this.currentRealScale = this.previewMode === G.Slide ? Math.min(t.width / n, t.height / i) : t.width / n, this.currentRealScale * this.currentScaleRatio;
  }
  processQueue(e) {
    const t = e[0];
    switch (t) {
      case "new":
      case "diff-v1":
        return t === "new" && this.reset(), this.kModule.manipulateData({
          action: "merge",
          data: e[1]
        }), this.moduleInitialized = !0, !0;
      case "viewport-change":
        return this.moduleInitialized ? !0 : (console.log("viewport-change before initialization"), !1);
      default:
        return console.log("svgUpdateEvent", e), !1;
    }
  }
  triggerUpdate() {
    if (this.isRendering)
      return;
    this.isRendering = !0;
    const e = async () => {
      if (this.cachedDOMState = this.retrieveDOMState(), this.patchQueue.length === 0) {
        this.isRendering = !1, this.postprocessChanges();
        return;
      }
      try {
        let t = performance.now();
        const n = this.canvasRenderCToken;
        n && (await n.cancel(), await n.wait(), this.canvasRenderCToken = void 0, console.log("cancel canvas rendering"));
        let i = !1;
        for (; this.patchQueue.length > 0; )
          i = this.processQueue(this.patchQueue.shift()) || i;
        let s = performance.now();
        i && (this.r.rescale(), await this.r.rerender(), this.r.rescale());
        let o = performance.now();
        const a = (d, g, f) => `${d} ${(f - g).toFixed(2)} ms`;
        this.sampledRenderTime = o - t, requestAnimationFrame(e);
      } catch (t) {
        console.error(t), this.isRendering = !1, this.postprocessChanges();
      }
    };
    requestAnimationFrame(e);
  }
  postprocessChanges() {
    this.r.postRender(), this.previewMode === G.Slide && document.querySelectorAll(".typst-page-number-indicator").forEach((e) => {
      e.textContent = `${this.kModule.retrievePagesInfo().length}`;
    });
  }
  addChangement(e) {
    e[0] === "new" && this.patchQueue.splice(0, this.patchQueue.length);
    const t = () => {
      this.vpTimeout = void 0, this.patchQueue.push(e), this.triggerUpdate();
    };
    this.vpTimeout !== void 0 && clearTimeout(this.vpTimeout), e[0] === "viewport-change" && this.isRendering ? this.vpTimeout = setTimeout(t, this.sampledRenderTime || 100) : t();
  }
  addViewportChange() {
    this.addChangement(["viewport-change", ""]);
  }
}
function Ir(r) {
  return class {
    constructor(t) {
      b(this, "impl");
      b(this, "kModule");
      if (t.isContentPreview && (t.renderMode = "canvas"), this.kModule = t.kModule, this.impl = new r(t), !this.impl.r)
        throw new Error(`mode is not supported, ${t == null ? void 0 : t.renderMode}`);
      t.isContentPreview && (this.impl.partialRendering = !0, this.impl.pixelPerPt = 1, this.impl.isMixinOutline = !0);
    }
    dispose() {
      this.impl.dispose();
    }
    reset() {
      this.impl.reset();
    }
    addChangement(t) {
      this.impl.addChangement(t);
    }
    addViewportChange() {
      this.impl.addViewportChange();
    }
    setPageColor(t) {
      this.impl.pageColor = t, this.addViewportChange();
    }
    setPartialRendering(t) {
      this.impl.partialRendering = t;
    }
    setCursor(t, n, i) {
      this.impl.cursorPosition = [t, n, i];
    }
    setPartialPageNumber(t) {
      return t <= 0 || t > this.kModule.retrievePagesInfo().length ? !1 : (this.impl.partialRenderPage = t - 1, this.addViewportChange(), !0);
    }
    getPartialPageNumber() {
      return this.impl.partialRenderPage + 1;
    }
    setOutineData(t) {
      this.impl.outline = t, this.addViewportChange();
    }
  };
}
function Rr(r, ...e) {
  return e.reduce((t, n) => n(t), r);
}
class Er {
  constructor() {
    b(this, "isCancellationRequested", !1);
    b(this, "_onCancelled");
    b(this, "_onCancelledResolveResolved");
    let e, t;
    this._onCancelled = new Promise((n) => {
      e = n, t && t(n);
    }), this._onCancelledResolveResolved = new Promise((n) => {
      t = n, e && n(e);
    });
  }
  async cancel() {
    await this._onCancelledResolveResolved, this.isCancellationRequested = !0;
  }
  isCancelRequested() {
    return this.isCancellationRequested;
  }
  async consume() {
    (await this._onCancelledResolveResolved)();
  }
  wait() {
    return this._onCancelled;
  }
}
const Cr = () => new Promise((r) => requestAnimationFrame(r));
var qe;
(function(r) {
  r[r.Doc = 0] = "Doc", r[r.Pages = 1] = "Pages";
})(qe || (qe = {}));
var H;
(function(r) {
  r[r.Layout = 0] = "Layout", r[r.Svg = 1] = "Svg", r[r.Semantics = 2] = "Semantics", r[r.PrepareCanvas = 3] = "PrepareCanvas", r[r.Canvas = 4] = "Canvas";
})(H || (H = {}));
function Ar(r) {
  return class extends r {
    constructor(...n) {
      super(...n);
      /// The template element for creating DOM by string.
      b(this, "tmpl", document.createElement("template"));
      /// The stub element for replacing an invisible element.
      b(this, "stub", this.createElement("<stub></stub>"));
      /// Typescript side of lib.
      b(this, "plugin");
      /// Rust side of kernel.
      b(this, "docKernel");
      /// The element to track.
      b(this, "resourceHeader");
      /// Expected exact state of the current DOM.
      /// Initially it is empty meaning no any page is rendered.
      b(this, "pages", []);
      /// The virtual scale of the document.
      b(this, "domScale", 1);
      /// Track mode.
      b(this, "track_mode", qe.Doc);
      /// Current executing task.
      b(this, "current_task");
      /// The currently maintained viewport.
      b(this, "viewport");
      if (this.registerMode("dom"), this.disposeList.push(() => {
        this.dispose();
      }), this.plugin = this.opts.renderer, this.opts.domScale !== void 0) {
        if (this.opts.domScale <= 0)
          throw new Error("domScale must be positive");
        this.domScale = this.opts.domScale;
      }
    }
    dispose() {
      for (const n of this.pages)
        n.dispose();
      this.docKernel && this.docKernel.free();
    }
    createElement(n) {
      return this.tmpl.innerHTML = n, this.tmpl.content.firstElementChild;
    }
    async mountDom(n) {
      if (this.docKernel)
        throw new Error("already mounted");
      this.hookedElem.innerHTML = '<svg class="typst-svg-resources" viewBox="0 0 0 0" width="0" height="0" style="opacity: 0; position: absolute;"></svg>', this.resourceHeader = this.hookedElem.querySelector(".typst-svg-resources"), this.docKernel = await this.plugin.renderer.mount_dom(this.kModule[S], this.hookedElem), this.docKernel.bind_functions({
        populateGlyphs: (i) => {
          let o = this.createElement(i).firstElementChild;
          this.resourceHeader.append(o);
        }
      });
    }
    async cancelAnyway$dom() {
      if (this.current_task) {
        const n = this.current_task;
        this.current_task = void 0, await n.cancel();
      }
    }
    retrieveDOMPages() {
      return Array.from(this.hookedElem.querySelectorAll(".typst-dom-page"));
    }
    // doesn't need to postRender
    postRender$dom() {
    }
    // doesn't need to rescale
    rescale$dom() {
    }
    getDomViewport(n, i) {
      const s = i.left, o = -i.top, a = i.right, d = n.innerHeight - i.top, g = {
        x: 0,
        y: o / this.domScale,
        width: Math.max(a - s, 0) / this.domScale,
        height: Math.max(d - o, 0) / this.domScale
      };
      return (g.width <= 0 || g.height <= 0) && (g.x = g.y = g.width = g.height = 0), g;
    }
    // fast mode
    async rerender$dom() {
      const n = this.retrieveDOMState(), { x: i, y: s, width: o, height: a } = this.getDomViewport(n.window, n.boundingRect);
      if (!await this.docKernel.relayout(i, s, o, a))
        return;
      const g = new Er();
      this.doRender$dom(g), this.current_task = g;
    }
    async doRender$dom(n) {
      const i = (d, g) => {
        if (d && !n.isCancelRequested() && g)
          return g();
      }, s = this.retrieveDOMPages().map((d) => {
        const { innerWidth: g, innerHeight: f } = window, m = d.getBoundingClientRect();
        return {
          inWindow: !(m.left > g || m.right < 0 || m.top > f || m.bottom < 0),
          page: d
        };
      }), o = async (d) => {
        if (await Cr(), n.isCancelRequested())
          return;
        const g = s[d].page, f = g.getBoundingClientRect(), m = this.getDomViewport(window, f), C = (B) => this.docKernel.need_repaint(d, m.x, m.y, m.width, m.height, B), E = (B) => this.docKernel.repaint(d, m.x, m.y, m.width, m.height, B), D = (B) => {
          if (!n.isCancelRequested())
            return i(C(B), () => E(B));
        };
        await D(H.Layout);
        const $ = (f.width ? Number.parseFloat(g.getAttribute("data-width")) / f.width : 1) * this.domScale, z = (f.height ? Number.parseFloat(g.getAttribute("data-height")) / f.height : 1) * this.domScale;
        m.x *= $, m.y *= z, m.y -= 100, m.width *= $, m.height *= z, m.height += 200, await D(H.Svg), await D(H.Semantics), !n.isCancelRequested() && (C(H.PrepareCanvas) ? (async () => {
          if (await E(H.PrepareCanvas), !n.isCancelRequested())
            return D(H.Canvas);
        })() : await D(H.Canvas));
      }, a = async (d) => {
        for (let g = 0; g < s.length; ++g) {
          if (n.isCancelRequested())
            return;
          s[g].inWindow === d && await o(g);
        }
      };
      this.cancelAnyway$dom(), await a(!0), await a(!1), n.isCancelRequested();
    }
  };
}
class Pr extends Ir(Rr(Ze, Ar)) {
}
var Ct;
let xe = (Ct = S, class {
  /**
   * @internal
   */
  constructor(e, t) {
    b(this, "plugin");
    /**
     * @internal
     */
    b(this, Ct);
    this.plugin = e, this[S] = t;
  }
  /**
   * @deprecated set in {@link RenderToCanvasOptions} instead
   *
   * Set the background color of the Typst document.
   * @param {string} t - The background color in format of `^#?[0-9a-f]{6}$`
   *
   * Note: Default to `#ffffff`.
   *
   * Note: Only available in canvas rendering mode.
   */
  set backgroundColor(e) {
    e !== void 0 && (this[S].background_color = e);
  }
  /**
   * Get the background color of the Typst document.
   *
   * Note: Default to `#ffffff`.
   *
   * Note: Only available in canvas rendering mode.
   */
  get backgroundColor() {
    return this[S].background_color;
  }
  /**
   * Set the pixel per point scale up the canvas panel.
   *
   * Note: Default to `3`.
   *
   * Note: Only available in canvas rendering mode.
   */
  set pixelPerPt(e) {
    e !== void 0 && (this[S].pixel_per_pt = e);
  }
  /**
   * @deprecated set in {@link RenderToCanvasOptions} instead
   *
   * Get the pixel per point scale up the canvas panel.
   *
   * Note: Default to `3`.
   *
   * Note: Only available in canvas rendering mode.
   */
  get pixelPerPt() {
    return this[S].pixel_per_pt;
  }
  /**
   * Reset state
   */
  reset() {
    this.plugin.resetSession(this);
  }
  /**
   * @deprecated
   * use {@link docWidth} instead
   */
  get doc_width() {
    return this[S].doc_width;
  }
  get docWidth() {
    return this[S].doc_width;
  }
  /**
   * @deprecated
   * use {@link docHeight} instead
   */
  get doc_height() {
    return this[S].doc_height;
  }
  get docHeight() {
    return this[S].doc_height;
  }
  retrievePagesInfo() {
    const e = this[S].pages_info, t = [], n = e.page_count;
    for (let i = 0; i < n; i++) {
      const s = e.page(i);
      t.push({
        pageOffset: s.page_off,
        width: s.width_pt,
        height: s.height_pt
      });
    }
    return t;
  }
  getSourceLoc(e) {
    return this[S].source_span(e);
  }
  /**
   * See {@link TypstRenderer#renderSvg} for more details.
   */
  renderSvg(e) {
    return this.plugin.renderSvg({
      renderSession: this,
      ...e
    });
  }
  /**
   * See {@link TypstRenderer#renderToSvg} for more details.
   */
  renderToSvg(e) {
    return this.plugin.renderToSvg({
      renderSession: this,
      ...e
    });
  }
  /**
   * See {@link TypstRenderer#renderCanvas} for more details.
   */
  renderCanvas(e) {
    return this.plugin.renderCanvas({
      renderSession: this,
      ...e
    });
  }
  /**
   * See {@link TypstRenderer#manipulateData} for more details.
   */
  manipulateData(e) {
    this.plugin.manipulateData({
      renderSession: this,
      ...e
    });
  }
  /**
   * See {@link TypstRenderer#renderSvgDiff} for more details.
   */
  renderSvgDiff(e) {
    return this.plugin.renderSvgDiff({
      renderSession: this,
      ...e
    });
  }
  /**
   * @deprecated
   * use {@link getSourceLoc} instead
   */
  get_source_loc(e) {
    return this[S].source_span(e);
  }
  /**
   * @deprecated
   * use {@link renderSvgDiff} instead
   */
  render_in_window(e, t, n, i) {
    return this[S].render_in_window(e, t, n, i);
  }
  /**
   * @deprecated
   * use {@link manipulateData} instead
   */
  merge_delta(e) {
    this.plugin.manipulateData({
      renderSession: this,
      action: "merge",
      data: e
    });
  }
});
var re;
(function(r) {
  r[r.Delete = 0] = "Delete", r[r.New = 1] = "New", r[r.Update = 2] = "Update";
})(re || (re = {}));
var At;
let Wt = (At = S, class {
  /**
   * @internal
   */
  constructor(e, t) {
    b(this, "plugin");
    /**
     * @internal
     */
    b(this, At);
    /**
     * @internal
     */
    b(this, "managedCanvasElemList", /* @__PURE__ */ new Map());
    /**
     * @internal
     */
    b(this, "canvasCounter", Math.random());
    this.plugin = e, this[S] = t;
  }
  /**
   * See {@link TypstRenderer#manipulateData} for more details.
   */
  manipulateData(e, t) {
    return this[S].manipulate_data(e, t);
  }
  /**
   * You must submit all canvas in pages to ensure synchronization with the background worker
   *
   * See {@link TypstRenderer#renderCanvas} for more details.
   */
  renderCanvas(e) {
    const t = this.managedCanvasElemList;
    for (const [a, d] of t)
      d[0] = re.Delete;
    for (const a of e) {
      const d = a.canvas;
      let g = d.dataset.manageId, f = re.Update;
      g || (g = this.canvasCounter.toFixed(5), this.canvasCounter += 1, d.dataset.manageId = g, f = re.New);
      let m = t.get(g);
      if (m && m[0] !== re.Delete)
        throw new Error("cannot update a canvas for two times in batch");
      t.set(g, [f, { ...a }]);
    }
    const n = Array.from(t.entries()), i = new Uint8Array(n.length), s = new Array(n.length), o = n.map(([a, [d, g]], f) => (d || t.delete(a), i[f] = d, s[f] = g.canvas, this.plugin.canvasOptionsToRust(g)));
    return this[S].render_canvas(i, s, o);
  }
  async retrievePagesInfo() {
    const e = await this[S].get_pages_info();
    console.log(e);
    const t = [], n = e.page_count;
    for (let i = 0; i < n; i++) {
      const s = e.page(i);
      t.push({
        pageOffset: s.page_off,
        width: s.width_pt,
        height: s.height_pt
      });
    }
    return t;
  }
});
const Fr = (r) => new Dt(async (e) => await r.default(e));
function zt() {
  return new Bt();
}
let rt = !0;
class Bt {
  constructor() {
    b(this, "renderer");
    b(this, "rendererJs");
  }
  async init(e) {
    var n;
    this.rendererJs = await (((n = e == null ? void 0 : e.getWrapper) == null ? void 0 : n.call(e)) || Promise.resolve().then(() => en));
    const t = this.rendererJs.TypstRendererBuilder;
    this.renderer = await $t(e, Fr(this.rendererJs), t, {});
  }
  loadGlyphPack(e) {
    return Promise.resolve();
  }
  createOptionsToRust(e) {
    const t = new this.rendererJs.CreateSessionOptions();
    return e.format !== void 0 && (t.format = e.format), e.artifactContent !== void 0 && (t.artifact_content = e.artifactContent), t;
  }
  canvasOptionsToRust(e) {
    const t = new this.rendererJs.RenderPageImageOptions();
    if (e.pageOffset === void 0)
      throw new Error("pageOffset is required in reflexo v0.5.0");
    if (t.page_off = e.pageOffset, e.cacheKey !== void 0 && (t.cache_key = e.cacheKey), e.backgroundColor !== void 0 && (t.background_color = e.backgroundColor), e.pixelPerPt !== void 0 && (t.pixel_per_pt = e.pixelPerPt), e.dataSelection !== void 0) {
      let n = 0;
      e.dataSelection.body ? n |= 1 : e.canvas && rt && (rt = !1, console.warn("dataSelection.body is not set but providing canvas for body")), (e.dataSelection.text || e.dataSelection.annotation) && console.error("dataSelection.text and dataSelection.annotation are deprecated"), e.dataSelection.semantics && (n |= 8), t.data_selection = n;
    }
    return t;
  }
  retrievePagesInfoFromSession(e) {
    return e.retrievePagesInfo();
  }
  /**
   * Render a Typst document to canvas.
   */
  renderCanvas(e) {
    return this.withinOptionSession(e, async (t) => this.renderer.render_page_to_canvas(t[S], e.canvas || void 0, this.canvasOptionsToRust(e)));
  }
  // async renderPdf(artifactContent: string): Promise<Uint8Array> {
  // return this.renderer.render_to_pdf(artifactContent);
  // }
  async inAnimationFrame(e) {
    return new Promise((t, n) => {
      requestAnimationFrame(() => {
        try {
          t(e());
        } catch (i) {
          n(i);
        }
      });
    });
  }
  async renderDisplayLayer(e, t, n) {
    const s = e[S].pages_info.page_count, o = async (f, m) => {
      const E = t[f].getContext("2d");
      if (!E)
        throw new Error("canvas context is null");
      return await this.renderCanvas({
        ...n,
        canvas: E,
        renderSession: e,
        pageOffset: m
      });
    }, a = performance.now(), d = await (async () => {
      const f = [];
      for (let m = 0; m < s; m++)
        f.push(await this.inAnimationFrame(() => o(m, m)));
      return f;
    })(), g = performance.now();
    return console.log(`display layer used: render = ${(g - a).toFixed(1)}ms`), d;
  }
  renderTextLayer(e, t) {
    const n = performance.now();
    e.forEach((s, o) => {
      s.innerHTML = t[o].htmlSemantics[0];
    });
    const i = performance.now();
    console.log(`text layer used: render = ${(i - n).toFixed(1)}ms`);
  }
  async render(e) {
    if ("format" in e && e.format !== "vector" && ["serde_json", "js", "ir"].includes(e.format))
      throw new Error(`deprecated format ${e.format}, please use vector format`);
    return this.renderToCanvas(e);
  }
  async renderDom(e) {
    if ("format" in e && e.format !== "vector" && ["serde_json", "js", "ir"].includes(e.format))
      throw new Error(`deprecated format ${e.format}, please use vector format`);
    return this.withinOptionSession(e, async (t) => {
      const n = new Pr({
        ...e,
        renderMode: "dom",
        hookedElem: e.container,
        kModule: t,
        renderer: this
      });
      return await n.impl.mountDom(e.pixelPerPt), n;
    });
  }
  async renderToCanvas(e) {
    let t, n;
    const i = e.container;
    i.style.visibility = "hidden";
    const s = async (o, a) => {
      try {
        n = await this.renderDisplayLayer(t, o, e), a();
      } finally {
        i.style.visibility = "visible";
      }
    };
    return this.withinOptionSession(e, async (o) => {
      if (t = o, t[S].pages_info.page_count === 0)
        throw new Error("No page found in session");
      if (e.pixelPerPt !== void 0 && e.pixelPerPt <= 0)
        throw new Error("Invalid typst.RenderOptions.pixelPerPt, should be a positive number " + e.pixelPerPt);
      let a = e.backgroundColor;
      if (a !== void 0 && !/^#[0-9a-f]{6}$/.test(a))
        throw new Error("Invalid typst.backgroundColor color for matching ^#?[0-9a-f]{6}$ " + a);
      t.pixelPerPt = e.pixelPerPt ?? Re.PIXEL_PER_PT, t.backgroundColor = a ?? "#ffffff";
      const d = performance.now(), g = new kr(this.retrievePagesInfoFromSession(t), i, e), f = performance.now();
      console.log(`layer used: retrieve = ${(f - d).toFixed(1)}ms`), await s(g.canvasList, () => g.resetLayout()), this.renderTextLayer(g.textLayerList, n);
    });
  }
  createModule(e) {
    return Promise.resolve(new xe(this, this.renderer.create_session(e && this.createOptionsToRust({
      format: "vector",
      artifactContent: e
    }))));
  }
  async createWorkerV0(e) {
    return new Wt(this, await this.renderer.create_worker(e));
  }
  workerBridge() {
    return this.renderer.create_worker_bridge();
  }
  renderSvg(e, t) {
    if (e instanceof xe || t)
      throw new Error("removed api, please use renderToSvg({ renderSession, container }) instead");
    return this.withinOptionSession(e, async (n) => {
      let i;
      return e.data_selection && (i = 0, e.data_selection.body && (i |= 1), e.data_selection.defs && (i |= 2), e.data_selection.css && (i |= 4), e.data_selection.js && (i |= 8)), Promise.resolve(this.renderer.svg_data(n[S], i));
    });
  }
  renderSvgDiff(e) {
    return e.window ? this.renderer.render_svg_diff(e.renderSession[S], e.window.lo.x, e.window.lo.y, e.window.hi.x, e.window.hi.y) : this.renderer.render_svg_diff(e.renderSession[S], 0, 0, 1e33, 1e33);
  }
  renderToSvg(e) {
    return this.withinOptionSession(e, async (t) => Promise.resolve(this.renderer.render_svg(t[S], e.container)));
  }
  getCustomV1(e) {
    return Promise.resolve(this.renderer.get_customs(e.renderSession[S]));
  }
  resetSession(e) {
    return this.renderer.reset(e[S]);
  }
  manipulateData(e) {
    return this.renderer.manipulate_data(e.renderSession[S], e.action ?? "reset", e.data);
  }
  withinOptionSession(e, t) {
    function n(i) {
      return "artifactContent" in i;
    }
    if ("renderSession" in e)
      return t(e.renderSession);
    if (n(e))
      return this.runWithSession(e, t);
    throw new Error("Invalid render options, should be one of RenderByContentOptions|RenderBySessionOptions");
  }
  async runWithSession(e, t) {
    let n = e, i = t;
    t || (n = void 0, i = e);
    const s = this.renderer.create_session(
      /* moved */
      n && this.createOptionsToRust(n)
    );
    try {
      const o = await i(new xe(this, s));
      return s.free(), o;
    } catch (o) {
      throw s.free(), o;
    }
  }
}
const nt = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  RenderSession: xe,
  TypstRendererDriver: Bt,
  TypstWorker: Wt,
  createTypstRenderer: zt
}, Symbol.toStringTag, { value: "Module" }));
class it {
  constructor() {
    b(this, "mTimes", /* @__PURE__ */ new Map());
    b(this, "mData", /* @__PURE__ */ new Map());
  }
  reset() {
    this.mTimes.clear(), this.mData.clear();
  }
  insertFile(e, t, n) {
    this.mTimes.set(e, n), this.mData.set(e, t);
  }
  removeFile(e) {
    this.mTimes.delete(e), this.mData.delete(e);
  }
  getMTime(e) {
    if (e.startsWith("/@memory/") && this.mTimes.has(e))
      return this.mTimes.get(e);
  }
  isFile() {
    return !0;
  }
  getRealPath(e) {
    return e;
  }
  readAll(e) {
    if (e.startsWith("/@memory/") && this.mData.has(e))
      return this.mData.get(e);
  }
}
class st {
  constructor(e) {
    b(this, "am");
    b(this, "cache", /* @__PURE__ */ new Map());
    this.am = e;
  }
  resolvePath(e) {
    return `https://packages.typst.org/preview/${e.name}-${e.version}.tar.gz`;
  }
  pullPackageData(e) {
    const t = new XMLHttpRequest();
    if (t.overrideMimeType("text/plain; charset=x-user-defined"), t.open("GET", this.resolvePath(e), !1), t.send(null), t.status === 200 && (t.response instanceof String || typeof t.response == "string"))
      return Uint8Array.from(t.response, (n) => n.charCodeAt(0));
  }
  resolve(e, t) {
    if (e.namespace !== "preview")
      return;
    const n = this.resolvePath(e);
    if (this.cache.has(n))
      return this.cache.get(n)();
    const i = this.pullPackageData(e);
    if (!i)
      return;
    const s = `/@memory/fetch/packages/preview/${e.namespace}/${e.name}/${e.version}`, o = [];
    t.untar(i, (d, g, f) => {
      o.push([s + "/" + d, g, new Date(f)]);
    });
    const a = () => {
      for (const [d, g, f] of o)
        this.am.insertFile(d, g, f);
      return s;
    };
    return this.cache.set(n, a), a();
  }
}
var ve;
(function(r) {
  r[r.vector = 0] = "vector", r[r.pdf = 1] = "pdf", r[r._dummy = 2] = "_dummy";
})(ve || (ve = {}));
var Pt;
Pt = S;
class Ut {
  /**
   * @internal
   */
  constructor(e) {
    /**
     * @internal
     */
    b(this, Pt);
    this[S] = e;
  }
  /**
   * Reset the incremental server to the initial state.
   */
  reset() {
    this[S].reset();
  }
  /**
   * Return current result.
   */
  current() {
    return this[S].current();
  }
  /**
   * Also attach the debug info to the result.
   */
  setAttachDebugInfo(e) {
    this[S].set_attach_debug_info(e);
  }
}
function Or() {
  return new Vt();
}
var Ft;
Ft = S;
class qt {
  constructor(e) {
    b(this, Ft);
    this[S] = e;
  }
  /**
   * Compile the paged document.
   *
   * @param {DiagnosticsFormat} format - The format of the diagnostics.
   * @returns {Promise<{ diagnostics?: DiagnosticsData[DiagnosticsFormat][] }>} - The result of the compilation.
   */
  compile(e) {
    return this[S].compile(0, ne(e == null ? void 0 : e.diagnostics));
  }
  /**
   * Compile the paged document.
   *
   * @param {DiagnosticsFormat} format - The format of the diagnostics.
   * @returns {Promise<{ diagnostics?: DiagnosticsData[DiagnosticsFormat][] }>} - The result of the compilation.
   */
  compileHtml(e) {
    return this[S].compile(1, ne(e == null ? void 0 : e.diagnostics));
  }
  /**
   * Runs query on the paged document.
   */
  async query(e) {
    return JSON.parse(this[S].query(0, e.selector, e.field));
  }
  /**
   * Get the title of the paged document.
   * Throw error if the world didn't compile the paged document.
   *
   * @returns {string | undefined} - The title of the paged document.
   */
  title() {
    return this[S].title(0);
  }
  /**
   * Export the paged document as vector format.
   *
   * @returns {Uint8Array | undefined} - The title of the paged document.
   */
  vector(e) {
    return this[S].get_artifact(0, ne(e == null ? void 0 : e.diagnostics)) || {};
  }
  /**
   * Export the paged document to PDF.
   *
   * @returns {Uint8Array | undefined} - The title of the paged document.
   */
  pdf(e) {
    return this[S].get_artifact(1, ne(e == null ? void 0 : e.diagnostics)) || {};
  }
}
const Nt = (r) => new Dt(async (e) => await r.default(e));
function Ht() {
  return new Ee();
}
class Vt {
  constructor() {
    b(this, "fontBuilderJs");
    b(this, "fontBuilder");
  }
  async init(e) {
    var t, n;
    this.fontBuilderJs = await (((t = e == null ? void 0 : e.getWrapper) == null ? void 0 : t.call(e)) || Promise.resolve().then(() => sr)), await Nt(this.fontBuilderJs).init((n = e == null ? void 0 : e.getModule) == null ? void 0 : n.call(e)), this.fontBuilder = new this.fontBuilderJs.TypstFontResolverBuilder();
  }
  async getFontInfo(e) {
    return this.fontBuilder.get_font_info(e);
  }
  async addFontData(e) {
    this.fontBuilder.add_raw_font(e);
  }
  async addLazyFont(e, t) {
    return this.fontBuilder.add_lazy_font(e, t);
  }
  async build(e) {
    const t = await this.fontBuilder.build(), n = await e(t);
    return t.free(), n;
  }
}
const Fe = class Fe {
  constructor() {
    b(this, "compiler");
    b(this, "compilerJs");
  }
  async init(e) {
    var g;
    this.compilerJs = await (((g = e == null ? void 0 : e.getWrapper) == null ? void 0 : g.call(e)) || Promise.resolve().then(() => sr));
    const t = this.compilerJs.TypstCompilerBuilder, n = { ...e || {} }, i = n.beforeBuild ?? (n.beforeBuild = []), s = i.some((f) => f._preloadRemoteFontOptions !== void 0), o = i.some((f) => {
      var m;
      return ((m = f._preloadRemoteFontOptions) == null ? void 0 : m.assets) !== void 0;
    }), a = i.some((f) => {
      var m;
      return ((m = f._preloadRemoteFontOptions) == null ? void 0 : m.assets) === !1;
    });
    if ((!s || !o && !a) && i.push(Oe([], { assets: Fe.defaultAssets })), !i.some((f) => f._kind === "fontLoader"))
      throw new Error("TypstCompiler: no font loader found, please use font loaders, e.g. loadFonts or preloadSystemFonts");
    this.compiler = await $t(e, Nt(this.compilerJs), t, {});
  }
  setFonts(e) {
    this.compiler.set_fonts(e);
  }
  compile(e) {
    return new Promise((t) => {
      const n = this.compiler.snapshot(e.root, e.mainFilePath, ot(e.inputs));
      if ("incrementalServer" in e) {
        t(n.incr_compile(e.incrementalServer[S], ne(e.diagnostics)));
        return;
      }
      t(n.get_artifact(e.format || ve.vector, ne(e.diagnostics)));
    });
  }
  async runWithWorld(e, t) {
    const n = this.compiler.snapshot(e.root, e.mainFilePath, ot(e.inputs));
    let i = await t(new qt(n));
    return n.free(), i;
  }
  query(e) {
    return this.runWithWorld(e, async (t) => JSON.parse(await t.query(e)));
  }
  getSemanticTokenLegend() {
    return new Promise((e) => {
      e(this.compiler.get_semantic_token_legend());
    });
  }
  getSemanticTokens(e) {
    return new Promise((t) => {
      this.compiler.reset(), t(this.compiler.get_semantic_tokens(e.offsetEncoding || "utf-16", e.mainFilePath, e.resultId));
    });
  }
  async withIncrementalServer(e) {
    const t = new Ut(this.compiler.create_incr_server());
    try {
      return await e(t);
    } finally {
      t[S].free();
    }
  }
  async getAst(e) {
    return this.compiler.get_ast(e);
  }
  async reset() {
    await new Promise((e) => {
      this.compiler.reset(), e(void 0);
    });
  }
  addSource(e, t) {
    if (arguments.length > 2)
      throw new Error("use of addSource(path, source, isMain) is deprecated, please use addSource(path, source) instead");
    this.compiler.add_source(e, t);
  }
  mapShadow(e, t) {
    this.compiler.map_shadow(e, t);
  }
  unmapShadow(e) {
    this.compiler.unmap_shadow(e);
  }
  resetShadow() {
    this.compiler.reset_shadow();
  }
  renderPageToCanvas() {
    throw new Error("Please use the api TypstRenderer.renderToCanvas in v0.4.0");
  }
};
b(Fe, "defaultAssets", ["text"]);
let Ee = Fe;
Ht._impl = Ee;
function ot(r) {
  return r ? Object.entries(r) : void 0;
}
function ne(r) {
  switch (r) {
    case "none":
      return 1;
    case "unix":
      return 2;
    case "full":
    default:
      return 3;
  }
}
const Me = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  get CompileFormatEnum() {
    return ve;
  },
  IncrementalServer: Ut,
  TypstFontBuilderDriver: Vt,
  TypstWorld: qt,
  createTypstCompiler: Ht,
  createTypstFontBuilder: Or
}, Symbol.toStringTag, { value: "Module" }));
function Lr(r) {
  return Math.random().toString(36).replace("0.", "");
}
const Tr = (
  // @ts-ignore
  typeof process < "u" && process.versions != null && process.versions.node != null
);
class U {
  /**
   * Create a new instance of {@link TypstSnippet}.
   * @param cc the compiler instance, see {@link PromiseJust} and {@link TypstCompiler}.
   * @param ex the renderer instance, see {@link PromiseJust} and {@link TypstRenderer}.
   *
   * @example
   *
   * Passes a global shared compiler instance that get initialized lazily:
   * ```typescript
   * const $typst = new TypstSnippet(() => {
   *  return createGlobalCompiler(createTypstCompiler, initOptions);
   * });
   *
   */
  constructor(e) {
    /** @internal */
    b(this, "mainFilePath");
    /** @internal */
    b(this, "cc");
    /** @internal */
    b(this, "fr");
    /** @internal */
    b(this, "ex");
    b(this, "providers");
    /** @internal */
    b(this, "ccOptions");
    /** @internal */
    b(this, "exOptions");
    b(this, "prepareUseOnce");
    this.cc = (e == null ? void 0 : e.compiler) || U.buildLocalCompiler, this.fr = (e == null ? void 0 : e.fontResolver) || U.buildLocalFontResolver, this.ex = (e == null ? void 0 : e.renderer) || U.buildLocalRenderer, this.mainFilePath = "/main.typ", this.providers = [];
  }
  /**
   * Set lazy initialized compiler instance for the utility instance.
   * @param cc the compiler instance, see {@link PromiseJust} and {@link TypstCompiler}.
   */
  setCompiler(e) {
    this.cc = e;
  }
  async getFontResolver() {
    return typeof this.fr == "function" ? this.fr = await this.fr() : this.fr;
  }
  /**
   * Get an initialized compiler instance from the utility instance.
   */
  async getCompiler() {
    return typeof this.cc == "function" ? this.cc = await this.cc() : this.cc;
  }
  async getCompilerReset() {
    const e = await this.getCompiler();
    return await e.reset(), e;
  }
  /**
   * Set lazy initialized renderer instance for the utility instance.
   * @param ex the renderer instance, see {@link PromiseJust} and {@link TypstRenderer}.
   */
  setRenderer(e) {
    this.ex = e;
  }
  /**
   * Get an initialized renderer instance from the utility instance.
   */
  async getRenderer() {
    return typeof this.ex == "function" ? this.ex = await this.ex() : this.ex;
  }
  /**
   * add providers for bullding the compiler or renderer component.
   */
  use(...e) {
    if (!this.providers)
      throw new Error("already prepare uses for instances");
    this.providers.push(...e);
  }
  /**
   * todo: add docs
   */
  static preloadFontFromUrl(e) {
    return U.preloadFonts([e]);
  }
  /**
   * todo: add docs
   */
  static preloadFontData(e) {
    return U.preloadFonts([e]);
  }
  /**
   * todo: add docs
   */
  static preloadFonts(e) {
    return {
      key: "access-model",
      forRoles: ["compiler"],
      provides: [Oe(e)]
    };
  }
  /**
   * don't load any default font assets.
   * todo: add docs
   */
  static disableDefaultFontAssets() {
    return {
      key: "access-model",
      forRoles: ["compiler"],
      provides: [mr()]
    };
  }
  /**
   * todo: add docs
   */
  static preloadFontAssets(e) {
    return {
      key: "access-model",
      forRoles: ["compiler"],
      provides: [yr(e)]
    };
  }
  /**
   * Set accessl model for the compiler instance
   * @example
   *
   * use memory access model
   *
   * ```typescript
   * const m = new MemoryAccessModel();
   * $typst.use(TypstSnippet.withAccessModel(m));
   * ```
   */
  static withAccessModel(e) {
    return {
      key: "access-model",
      forRoles: ["compiler"],
      provides: [tt(e)]
    };
  }
  /**
   * Set package registry for the compiler instance
   * @example
   *
   * use a customized package registry
   *
   * ```typescript
   * const n = new NodeFetchPackageRegistry();
   * $typst.use(TypstSnippet.withPackageRegistry(n));
   * ```
   */
  static withPackageRegistry(e) {
    return {
      key: "package-registry",
      forRoles: ["compiler"],
      provides: [Te(e)]
    };
  }
  /**
   * Retrieve an access model to store the data of fetched files.
   * Provide a PackageRegistry instance for the compiler instance.
   *
   * @example
   *
   * use default (memory) access model
   *
   * ```typescript
   * $typst.use(await TypstSnippet.fetchPackageRegistry());
   * ```
   *
   * @example
   *
   * use external access model
   *
   * ```typescript
   * const m = new MemoryAccessModel();
   * $typst.use(TypstSnippet.withAccessModel(m), await TypstSnippet.fetchPackageRegistry(m));
   * ```
   */
  static fetchPackageRegistry(e) {
    const t = e || new it(), n = [
      ...e ? [] : [tt(t)],
      Te(new st(t))
    ];
    return {
      key: "package-registry$fetch",
      forRoles: ["compiler"],
      provides: n
    };
  }
  /**
   * Retrieve a fetcher for fetching package data.
   * Provide a PackageRegistry instance for the compiler instance.
   * @example
   *
   * use a customized fetcher
   *
   * ```typescript
   * import request from 'sync-request-curl';
   * const m = new MemoryAccessModel();
   * $typst.use(TypstSnippet.withAccessModel(m), await TypstSnippet.fetchPackageBy(m, (_, httpUrl) => {
   *   const response = request('GET', this.resolvePath(path), {
   *     insecure: true,
   *   });
   *
   *   if (response.statusCode === 200) {
   *     return response.getBody(undefined);
   *   }
   *   return undefined;
   * }));
   * ```
   */
  static fetchPackageBy(e, t) {
    class n extends st {
      pullPackageData(s) {
        return t(s, this.resolvePath(s));
      }
    }
    return {
      key: "package-registry$lambda",
      forRoles: ["compiler"],
      provides: [Te(new n(e))]
    };
  }
  /**
   * Set compiler init options for initializing global instance {@link $typst}.
   * See {@link InitOptions}.
   */
  setCompilerInitOptions(e) {
    this.requireIsUninitialized("compiler", this.cc), this.ccOptions = e;
  }
  /**
   * Set renderer init options for initializing global instance {@link $typst}.
   * See {@link InitOptions}.
   */
  setRendererInitOptions(e) {
    this.requireIsUninitialized("renderer", this.ex), this.exOptions = e;
  }
  /**
   * Set shared main file path.
   */
  setMainFilePath(e) {
    this.mainFilePath = e;
  }
  /**
   * Get shared main file path.
   */
  getMainFilePath() {
    return this.mainFilePath;
  }
  removeTmp(e) {
    return e.mainFilePath.startsWith("/tmp/") ? this.unmapShadow(e.mainFilePath) : Promise.resolve();
  }
  /**
   * Adds a font to the compiler.
   *
   * @example
   *
   * ```typescript
   * const fonts = await fetch('fontInfo.json').then(res => res.json());
   * $typst.addFonts(fonts.map(font => $typst.loadFont(font.url)));
   * ```
   *
   * @param fontInfos the font infos to add.
   */
  async setFonts(e) {
    const t = await this.getFontResolver();
    for (const i of e)
      await t.addLazyFont(i, "blob" in i ? i.blob : jt(i), i);
    const n = await this.getCompiler();
    await t.build(async (i) => n.setFonts(i));
  }
  /**
   * Add a source file to the compiler.
   * See {@link TypstCompiler#addSource}.
   */
  async addSource(e, t) {
    (await this.getCompiler()).addSource(e, t);
  }
  /**
   * Reset the shadow files.
   * Note: this function is independent to the {@link reset} function.
   * See {@link TypstCompiler#resetShadow}.
   */
  async resetShadow() {
    (await this.getCompiler()).resetShadow();
  }
  /**
   * Add a shadow file to the compiler.
   * See {@link TypstCompiler#mapShadow}.
   */
  async mapShadow(e, t) {
    (await this.getCompiler()).mapShadow(e, t);
  }
  /**
   * Remove a shadow file from the compiler.
   * See {@link TypstCompiler#unmapShadow}.
   */
  async unmapShadow(e) {
    (await this.getCompiler()).unmapShadow(e);
  }
  /**
   * Compile the document to vector (IR) format.
   * See {@link SweetCompileOptions}.
   */
  async vector(e) {
    const t = await this.getCompileOptions(e);
    return (await this.getCompilerReset()).compile(t).then((i) => i.result).finally(() => this.removeTmp(t));
  }
  /**
   * Compile the document to PDF format.
   * See {@link SweetCompileOptions}.
   */
  async pdf(e) {
    const t = await this.getCompileOptions(e);
    return t.format = ve.pdf, (await this.getCompilerReset()).compile(t).then((i) => i.result).finally(() => this.removeTmp(t));
  }
  /**
   * Compile the document to SVG format.
   * See {@link SweetRenderOptions} and {@link RenderSvgOptions}.
   */
  async svg(e) {
    return this.transientRender(e, (t, n) => t.renderSvg({
      ...e,
      renderSession: n
    }));
  }
  /**
   * Compile the document to canvas operations.
   * See {@link SweetRenderOptions} and {@link RenderToCanvasOptions}.
   */
  async canvas(e, t) {
    return this.transientRender(t, (n, i) => n.renderToCanvas({
      container: e,
      ...t,
      renderSession: i
    }));
  }
  /**
   * Get semantic tokens for the document.
   */
  async query(e) {
    const t = await this.getCompileOptions(e);
    return (await this.getCompilerReset()).query({
      ...e,
      ...t
    }).finally(() => this.removeTmp(t));
  }
  /**
   * Get token legend for semantic tokens.
   */
  async getSemanticTokenLegend() {
    return (await this.getCompilerReset()).getSemanticTokenLegend();
  }
  /**
   * Get semantic tokens for the document.
   * See {@link SweetCompileOptions}.
   * See {@link TypstCompiler#getSemanticTokens}.
   */
  async getSemanticTokens(e) {
    const t = await this.getCompileOptions(e);
    return (await this.getCompilerReset()).getSemanticTokens({
      mainFilePath: t.mainFilePath,
      resultId: e.resultId
    }).finally(() => this.removeTmp(t));
  }
  async getCompileOptions(e) {
    if (e === void 0)
      return { mainFilePath: this.mainFilePath, diagnostics: "none" };
    if (typeof e == "string")
      throw new Error("please specify opts as {mainContent: '...'} or {mainFilePath: '...'}");
    if ("mainFilePath" in e)
      return { ...e, diagnostics: "none" };
    {
      const t = `/tmp/${Lr()}.typ`;
      return await this.addSource(t, e.mainContent), { mainFilePath: t, inputs: e.inputs, diagnostics: "none" };
    }
  }
  async getVector(e) {
    if (e && "vectorData" in e)
      return e.vectorData;
    const t = await this.getCompileOptions(e);
    return (await this.getCompiler()).compile(t).then((n) => n.result).finally(() => this.removeTmp(t));
  }
  async transientRender(e, t) {
    const n = await this.getRenderer();
    if (!n)
      throw new Error("does not provide renderer instance");
    const i = await this.getVector(e);
    return await n.runWithSession(async (s) => (n.manipulateData({
      renderSession: s,
      action: "reset",
      data: i
    }), t(n, s)));
  }
  async prepareUse() {
    return this.prepareUseOnce ? this.prepareUseOnce : this.prepareUseOnce = this.doPrepareUse();
  }
  async doPrepareUse() {
    if (!this.providers)
      return;
    const e = await Promise.all(this.providers.map((a) => typeof a == "function" ? a() : a));
    if (this.providers = [], De == this && !e.some((a) => a.key.includes("package-registry") || a.key.includes("access-model")))
      if (Tr) {
        const a = new Function("m", "return import(m)");
        try {
          const d = new it(), { default: g } = await a("sync-request");
          De.use(U.withAccessModel(d), U.fetchPackageBy(d, (f, m) => {
            const C = g("GET", m);
            if (C.statusCode === 200)
              return C.getBody(void 0);
          }));
        } catch {
        }
      } else
        De.use(U.fetchPackageRegistry());
    const t = await Promise.all(this.providers.map((a) => typeof a == "function" ? a() : a)), n = this.ccOptions || (this.ccOptions = {}), i = n.beforeBuild || (n.beforeBuild = []), s = this.exOptions || (this.exOptions = {}), o = s.beforeBuild || (s.beforeBuild = []);
    for (const a of [...e, ...t])
      a.forRoles.includes("compiler") && (this.requireIsUninitialized("compiler", this.cc), i.push(...a.provides)), a.forRoles.includes("renderer") && (this.requireIsUninitialized("renderer", this.ex), o.push(...a.provides));
    this.providers = void 0;
  }
  requireIsUninitialized(e, t, n) {
    if (t && typeof t != "function")
      throw new Error(`${e} has been initialized: ${t}`);
  }
  /** @internal */
  static async buildLocalCompiler() {
    const { createTypstCompiler: e } = await Promise.resolve().then(() => Me);
    await this.prepareUse();
    const t = e();
    return await t.init(this.ccOptions), t;
  }
  /** @internal */
  static async buildLocalFontResolver() {
    const { createTypstFontBuilder: e } = await Promise.resolve().then(() => Me);
    await this.prepareUse();
    const t = e();
    return await t.init(this.ccOptions), t;
  }
  /** @internal */
  static async buildGlobalCompiler() {
    const { createGlobalCompiler: e } = await Promise.resolve().then(() => hn), { createTypstCompiler: t } = await Promise.resolve().then(() => Me);
    return await this.prepareUse(), e(t, this.ccOptions);
  }
  /** @internal */
  static async buildLocalRenderer() {
    const { createTypstRenderer: e } = await Promise.resolve().then(() => nt);
    await this.prepareUse();
    const t = e();
    return await t.init(this.exOptions), t;
  }
  /** @internal */
  static async buildGlobalRenderer() {
    const { createGlobalRenderer: e } = await Promise.resolve().then(() => wr), { createTypstRenderer: t } = await Promise.resolve().then(() => nt);
    return await this.prepareUse(), e(t, this.exOptions);
  }
}
const De = new U({
  compiler: U.buildGlobalCompiler,
  renderer: U.buildGlobalRenderer
}), Mr = ".typst-html-semantics{position:absolute;z-index:2;color:transparent;font-family:monospace;white-space:pre}.typst-html-semantics span{color:transparent;font-family:monospace;transform-origin:left top;position:absolute;display:inline-block;left:0;top:0}.typst-content-hint{position:absolute;display:inline-block;width:1px;height:1px;overflow:hidden}.typst-html-semantics a{position:absolute;display:inline-block}.typst-content-group{pointer-events:visible}.typst-html-semantics span::-moz-selection{color:transparent;background:#7db9dea0}.typst-html-semantics span::selection{color:transparent;background:#7db9dea0}.typst-html-semantics *::-moz-selection{color:transparent;background:transparent}.typst-html-semantics *::selection{color:transparent;background:transparent}.typst-content-fallback{color:transparent;background:transparent}.pseudo-link,.typst-text{pointer-events:none}";
let Jt = {
  beforeBuild: [],
  getModule: () => "/node_modules/@myriaddreamin/typst-ts-renderer/pkg/typst_ts_renderer_bg.wasm"
};
const Dr = ({ fill: r, artifact: e, format: t }) => {
  const n = lr(null), i = () => n == null ? void 0 : n.current;
  return fr(() => {
    const s = (a) => {
      const d = i();
      if (d)
        return a.render({
          artifactContent: e,
          format: "vector",
          backgroundColor: r,
          container: d,
          pixelPerPt: 3
        });
    }, o = i();
    if (o) {
      if (!(e != null && e.length)) {
        o.innerHTML = "";
        return;
      }
      Mt(zt, Jt, s);
    }
  }, [n, r, e, t]), /* @__PURE__ */ gr("div", { children: [
    /* @__PURE__ */ et("style", { children: Mr }),
    /* @__PURE__ */ et("div", { className: "typst-app", style: { height: "0" }, ref: n })
  ] });
};
Dr.setWasmModuleInitOptions = (r) => {
  Jt = r;
};
let _, V = new Array(128).fill(void 0);
V.push(void 0, null, !0, !1);
function u(r) {
  return V[r];
}
let ie = V.length;
function v(r) {
  ie === V.length && V.push(V.length + 1);
  const e = ie;
  if (ie = V[e], typeof ie != "number") throw new Error("corrupt heap");
  return V[e] = r, e;
}
function I(r, e) {
  try {
    return r.apply(this, e);
  } catch (t) {
    _.__wbindgen_export_0(v(t));
  }
}
function l(r, e) {
  try {
    return r.apply(this, e);
  } catch (t) {
    let n = function() {
      try {
        return t instanceof Error ? `${t.message}

Stack:
${t.stack}` : t.toString();
      } catch {
        return "<failed to stringify thrown value>";
      }
    }();
    throw console.error("wasm-bindgen: imported JS function that was not marked as `catch` threw an error:", n), t;
  }
}
function p(r) {
  if (typeof r != "number") throw new Error(`expected a number argument, found ${typeof r}`);
}
let fe = null;
function se() {
  return (fe === null || fe.byteLength === 0) && (fe = new Uint8Array(_.memory.buffer)), fe;
}
let Se = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 });
Se.decode();
const $r = 2146435072;
let $e = 0;
function jr(r, e) {
  return $e += e, $e >= $r && (Se = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 }), Se.decode(), $e = e), Se.decode(se().subarray(r, r + e));
}
function R(r, e) {
  return r = r >>> 0, jr(r, e);
}
let O = 0;
const he = new TextEncoder();
"encodeInto" in he || (he.encodeInto = function(r, e) {
  const t = he.encode(r);
  return e.set(t), {
    read: r.length,
    written: t.length
  };
});
function j(r, e, t) {
  if (typeof r != "string") throw new Error(`expected a string argument, found ${typeof r}`);
  if (t === void 0) {
    const a = he.encode(r), d = e(a.length, 1) >>> 0;
    return se().subarray(d, d + a.length).set(a), O = a.length, d;
  }
  let n = r.length, i = e(n, 1) >>> 0;
  const s = se();
  let o = 0;
  for (; o < n; o++) {
    const a = r.charCodeAt(o);
    if (a > 127) break;
    s[i + o] = a;
  }
  if (o !== n) {
    o !== 0 && (r = r.slice(o)), i = t(i, n, n = o + r.length * 3, 1) >>> 0;
    const a = se().subarray(i + o, i + n), d = he.encodeInto(r, a);
    if (d.read !== r.length) throw new Error("failed to pass whole string");
    o += d.written, i = t(i, n, o, 1) >>> 0;
  }
  return O = o, i;
}
let X = null;
function h() {
  return (X === null || X.buffer.detached === !0 || X.buffer.detached === void 0 && X.buffer !== _.memory.buffer) && (X = new DataView(_.memory.buffer)), X;
}
function F(r) {
  return r == null;
}
function M(r) {
  if (typeof r != "boolean")
    throw new Error(`expected a boolean argument, found ${typeof r}`);
}
function _t(r, e) {
  return r = r >>> 0, se().subarray(r / 1, r / 1 + e);
}
function Wr(r) {
  r < 132 || (V[r] = ie, ie = r);
}
function P(r) {
  const e = u(r);
  return Wr(r), e;
}
function Ne(r) {
  const e = typeof r;
  if (e == "number" || e == "boolean" || r == null)
    return `${r}`;
  if (e == "string")
    return `"${r}"`;
  if (e == "symbol") {
    const i = r.description;
    return i == null ? "Symbol" : `Symbol(${i})`;
  }
  if (e == "function") {
    const i = r.name;
    return typeof i == "string" && i.length > 0 ? `Function(${i})` : "Function";
  }
  if (Array.isArray(r)) {
    const i = r.length;
    let s = "[";
    i > 0 && (s += Ne(r[0]));
    for (let o = 1; o < i; o++)
      s += ", " + Ne(r[o]);
    return s += "]", s;
  }
  const t = /\[object ([^\]]+)\]/.exec(toString.call(r));
  let n;
  if (t && t.length > 1)
    n = t[1];
  else
    return toString.call(r);
  if (n == "Object")
    try {
      return "Object(" + JSON.stringify(r) + ")";
    } catch {
      return "Object";
    }
  return r instanceof Error ? `${r.name}: ${r.message}
${r.stack}` : n;
}
const Ce = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry(
  (r) => {
    _.__wbindgen_export_4.get(r.dtor)(r.a, r.b);
  }
);
function at(r, e, t, n) {
  const i = { a: r, b: e, cnt: 1, dtor: t }, s = (...o) => {
    i.cnt++;
    try {
      return n(i.a, i.b, ...o);
    } finally {
      --i.cnt === 0 && (_.__wbindgen_export_4.get(i.dtor)(i.a, i.b), i.a = 0, Ce.unregister(i));
    }
  };
  return s.original = i, Ce.register(s, i, i), s;
}
function zr(r, e, t, n) {
  const i = { a: r, b: e, cnt: 1, dtor: t }, s = (...o) => {
    i.cnt++;
    const a = i.a;
    i.a = 0;
    try {
      return n(a, i.b, ...o);
    } finally {
      --i.cnt === 0 ? (_.__wbindgen_export_4.get(i.dtor)(a, i.b), Ce.unregister(i)) : i.a = a;
    }
  };
  return s.original = i, Ce.register(s, i, i), s;
}
function Br() {
  const r = _.renderer_build_info();
  return P(r);
}
function N(r, e) {
  if (!(r instanceof e))
    throw new Error(`expected instance of ${e.name}`);
}
function Ae(r, e) {
  const t = e(r.length * 1, 1) >>> 0;
  return se().set(r, t / 1), O = r.length, t;
}
let we = null;
function Ur() {
  return (we === null || we.byteLength === 0) && (we = new Uint32Array(_.memory.buffer)), we;
}
function qr(r, e) {
  const t = e(r.length * 4, 4) >>> 0;
  return Ur().set(r, t / 4), O = r.length, t;
}
function ct(r, e) {
  const t = e(r.length * 4, 4) >>> 0, n = h();
  for (let i = 0; i < r.length; i++)
    n.setUint32(t + 4 * i, v(r[i]), !0);
  return O = r.length, t;
}
function Nr(r, e, t) {
  p(r), p(e), _.__wbindgen_export_5(r, e, v(t));
}
function Hr(r, e) {
  p(r), p(e), _.__wbindgen_export_6(r, e);
}
function Vr(r, e, t) {
  p(r), p(e), _.__wbindgen_export_7(r, e, v(t));
}
function Jr(r, e, t, n) {
  p(r), p(e), _.__wbindgen_export_8(r, e, v(t), v(n));
}
const ut = ["nonzero", "evenodd"], dt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_createsessionoptions_free(r >>> 0, 1));
class Pe {
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, dt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_createsessionoptions_free(e, 0);
  }
  constructor() {
    const e = _.createsessionoptions_new();
    return this.__wbg_ptr = e >>> 0, dt.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @param {string} format
   */
  set format(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const t = j(e, _.__wbindgen_export_1, _.__wbindgen_export_2), n = O;
    _.createsessionoptions_set_format(this.__wbg_ptr, t, n);
  }
  /**
   * @param {Uint8Array} artifact_content
   */
  set artifact_content(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const t = Ae(e, _.__wbindgen_export_1), n = O;
    _.createsessionoptions_set_artifact_content(this.__wbg_ptr, t, n);
  }
}
Symbol.dispose && (Pe.prototype[Symbol.dispose] = Pe.prototype.free);
const gt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_incrdomdocclient_free(r >>> 0, 1));
class _e {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(_e.prototype);
    return t.__wbg_ptr = e, gt.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, gt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_incrdomdocclient_free(e, 0);
  }
  /**
   * @param {any} functions
   */
  bind_functions(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), _.incrdomdocclient_bind_functions(this.__wbg_ptr, v(e));
  }
  /**
   * Relayout the document in the given window.
   * @param {number} x
   * @param {number} y
   * @param {number} w
   * @param {number} h
   * @returns {Promise<boolean>}
   */
  relayout(e, t, n, i) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const s = _.incrdomdocclient_relayout(this.__wbg_ptr, e, t, n, i);
    return P(s);
  }
  /**
   * @param {number} page_num
   * @param {number} x
   * @param {number} y
   * @param {number} w
   * @param {number} h
   * @param {number} stage
   * @returns {boolean}
   */
  need_repaint(e, t, n, i, s, o) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const f = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), p(e), p(o), _.incrdomdocclient_need_repaint(f, this.__wbg_ptr, e, t, n, i, s, o);
      var a = h().getInt32(f + 0, !0), d = h().getInt32(f + 4, !0), g = h().getInt32(f + 8, !0);
      if (g)
        throw P(d);
      return a !== 0;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} page_num
   * @param {number} x
   * @param {number} y
   * @param {number} w
   * @param {number} h
   * @param {number} stage
   * @returns {any}
   */
  repaint(e, t, n, i, s, o) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const f = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), p(e), p(o), _.incrdomdocclient_repaint(f, this.__wbg_ptr, e, t, n, i, s, o);
      var a = h().getInt32(f + 0, !0), d = h().getInt32(f + 4, !0), g = h().getInt32(f + 8, !0);
      if (g)
        throw P(d);
      return P(a);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (_e.prototype[Symbol.dispose] = _e.prototype.free);
const lt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_pageinfo_free(r >>> 0, 1));
class Y {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(Y.prototype);
    return t.__wbg_ptr = e, lt.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, lt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_pageinfo_free(e, 0);
  }
  /**
   * @returns {number}
   */
  get page_off() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.pageinfo_page_off(this.__wbg_ptr) >>> 0;
  }
  /**
   * @returns {number}
   */
  get width_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.pageinfo_width_pt(this.__wbg_ptr);
  }
  /**
   * @returns {number}
   */
  get height_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.pageinfo_height_pt(this.__wbg_ptr);
  }
}
Symbol.dispose && (Y.prototype[Symbol.dispose] = Y.prototype.free);
const ft = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_pagesinfo_free(r >>> 0, 1));
class ae {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(ae.prototype);
    return t.__wbg_ptr = e, ft.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ft.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_pagesinfo_free(e, 0);
  }
  /**
   * @returns {number}
   */
  get page_count() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.pagesinfo_page_count(this.__wbg_ptr) >>> 0;
  }
  /**
   * @param {number} num
   * @returns {PageInfo | undefined}
   */
  page_by_number(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), p(e);
    const t = _.pagesinfo_page_by_number(this.__wbg_ptr, e);
    return t === 0 ? void 0 : Y.__wrap(t);
  }
  /**
   * @param {number} i
   * @returns {PageInfo}
   */
  page(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), p(e);
    const t = _.pagesinfo_page(this.__wbg_ptr, e);
    return Y.__wrap(t);
  }
  /**
   * @returns {number}
   */
  width() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.pagesinfo_width(this.__wbg_ptr);
  }
  /**
   * @returns {number}
   */
  height() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.pagesinfo_height(this.__wbg_ptr);
  }
}
Symbol.dispose && (ae.prototype[Symbol.dispose] = ae.prototype.free);
const wt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_renderpageimageoptions_free(r >>> 0, 1));
class Z {
  static __unwrap(e) {
    return e instanceof Z ? e.__destroy_into_raw() : 0;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, wt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_renderpageimageoptions_free(e, 0);
  }
  constructor() {
    const e = _.renderpageimageoptions_new();
    return this.__wbg_ptr = e >>> 0, wt.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @returns {number | undefined}
   */
  get pixel_per_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const e = _.renderpageimageoptions_pixel_per_pt(this.__wbg_ptr);
    return e === 4294967297 ? void 0 : e;
  }
  /**
   * @param {number | null} [pixel_per_pt]
   */
  set pixel_per_pt(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), F(e) || p(e), _.renderpageimageoptions_set_pixel_per_pt(this.__wbg_ptr, F(e) ? 4294967297 : Math.fround(e));
  }
  /**
   * @returns {string | undefined}
   */
  get background_color() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), _.renderpageimageoptions_background_color(n, this.__wbg_ptr);
      var e = h().getInt32(n + 0, !0), t = h().getInt32(n + 4, !0);
      let i;
      return e !== 0 && (i = R(e, t).slice(), _.__wbindgen_export_3(e, t * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [background_color]
   */
  set background_color(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    var t = F(e) ? 0 : j(e, _.__wbindgen_export_1, _.__wbindgen_export_2), n = O;
    _.renderpageimageoptions_set_background_color(this.__wbg_ptr, t, n);
  }
  /**
   * @returns {number}
   */
  get page_off() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.renderpageimageoptions_page_off(this.__wbg_ptr) >>> 0;
  }
  /**
   * @param {number} page_off
   */
  set page_off(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), p(e), _.renderpageimageoptions_set_page_off(this.__wbg_ptr, e);
  }
  /**
   * @returns {string | undefined}
   */
  get cache_key() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), _.renderpageimageoptions_cache_key(n, this.__wbg_ptr);
      var e = h().getInt32(n + 0, !0), t = h().getInt32(n + 4, !0);
      let i;
      return e !== 0 && (i = R(e, t).slice(), _.__wbindgen_export_3(e, t * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [cache_key]
   */
  set cache_key(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    var t = F(e) ? 0 : j(e, _.__wbindgen_export_1, _.__wbindgen_export_2), n = O;
    _.renderpageimageoptions_set_cache_key(this.__wbg_ptr, t, n);
  }
  /**
   * @returns {number | undefined}
   */
  get data_selection() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const e = _.renderpageimageoptions_data_selection(this.__wbg_ptr);
    return e === 4294967297 ? void 0 : e;
  }
  /**
   * @param {number | null} [data_selection]
   */
  set data_selection(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), F(e) || p(e), _.renderpageimageoptions_set_data_selection(this.__wbg_ptr, F(e) ? 4294967297 : e >>> 0);
  }
}
Symbol.dispose && (Z.prototype[Symbol.dispose] = Z.prototype.free);
const pt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_rendersession_free(r >>> 0, 1));
class W {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(W.prototype);
    return t.__wbg_ptr = e, pt.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, pt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_rendersession_free(e, 0);
  }
  /**
   * @returns {number | undefined}
   */
  get pixel_per_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const e = _.rendersession_pixel_per_pt(this.__wbg_ptr);
    return e === 4294967297 ? void 0 : e;
  }
  /**
   * @param {number | null} [pixel_per_pt]
   */
  set pixel_per_pt(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), F(e) || p(e), _.rendersession_set_pixel_per_pt(this.__wbg_ptr, F(e) ? 4294967297 : Math.fround(e));
  }
  /**
   * @returns {string | undefined}
   */
  get background_color() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), _.rendersession_background_color(n, this.__wbg_ptr);
      var e = h().getInt32(n + 0, !0), t = h().getInt32(n + 4, !0);
      let i;
      return e !== 0 && (i = R(e, t).slice(), _.__wbindgen_export_3(e, t * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [background_color]
   */
  set background_color(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    var t = F(e) ? 0 : j(e, _.__wbindgen_export_1, _.__wbindgen_export_2), n = O;
    _.rendersession_set_background_color(this.__wbg_ptr, t, n);
  }
  /**
   * @returns {PagesInfo}
   */
  get pages_info() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const e = _.rendersession_pages_info(this.__wbg_ptr);
    return ae.__wrap(e);
  }
  /**
   * @returns {number}
   */
  get doc_width() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.rendersession_doc_width(this.__wbg_ptr);
  }
  /**
   * @returns {number}
   */
  get doc_height() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    return p(this.__wbg_ptr), _.rendersession_doc_height(this.__wbg_ptr);
  }
  /**
   * @param {Uint32Array} path
   * @returns {string | undefined}
   */
  source_span(e) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr);
      const a = qr(e, _.__wbindgen_export_1), d = O;
      _.rendersession_source_span(o, this.__wbg_ptr, a, d);
      var t = h().getInt32(o + 0, !0), n = h().getInt32(o + 4, !0), i = h().getInt32(o + 8, !0), s = h().getInt32(o + 12, !0);
      if (s)
        throw P(i);
      let g;
      return t !== 0 && (g = R(t, n).slice(), _.__wbindgen_export_3(t, n * 1, 1)), g;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} rect_lo_x
   * @param {number} rect_lo_y
   * @param {number} rect_hi_x
   * @param {number} rect_hi_y
   * @returns {string}
   */
  render_in_window(e, t, n, i) {
    let s, o;
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const g = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), _.rendersession_render_in_window(g, this.__wbg_ptr, e, t, n, i);
      var a = h().getInt32(g + 0, !0), d = h().getInt32(g + 4, !0);
      return s = a, o = d, R(a, d);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16), _.__wbindgen_export_3(s, o, 1);
    }
  }
}
Symbol.dispose && (W.prototype[Symbol.dispose] = W.prototype.free);
const bt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_rendersessionoptions_free(r >>> 0, 1));
class He {
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, bt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_rendersessionoptions_free(e, 0);
  }
  constructor() {
    const e = _.rendersessionoptions_new();
    return this.__wbg_ptr = e >>> 0, bt.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @returns {number | undefined}
   */
  get pixel_per_pt() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const e = _.rendersession_pixel_per_pt(this.__wbg_ptr);
    return e === 4294967297 ? void 0 : e;
  }
  /**
   * @param {number | null} [pixel_per_pt]
   */
  set pixel_per_pt(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr), F(e) || p(e), _.rendersession_set_pixel_per_pt(this.__wbg_ptr, F(e) ? 4294967297 : Math.fround(e));
  }
  /**
   * @returns {string | undefined}
   */
  get background_color() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), _.rendersessionoptions_background_color(n, this.__wbg_ptr);
      var e = h().getInt32(n + 0, !0), t = h().getInt32(n + 4, !0);
      let i;
      return e !== 0 && (i = R(e, t).slice(), _.__wbindgen_export_3(e, t * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [background_color]
   */
  set background_color(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    var t = F(e) ? 0 : j(e, _.__wbindgen_export_1, _.__wbindgen_export_2), n = O;
    _.rendersessionoptions_set_background_color(this.__wbg_ptr, t, n);
  }
  /**
   * @returns {string | undefined}
   */
  get format() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const n = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), _.rendersessionoptions_format(n, this.__wbg_ptr);
      var e = h().getInt32(n + 0, !0), t = h().getInt32(n + 4, !0);
      let i;
      return e !== 0 && (i = R(e, t).slice(), _.__wbindgen_export_3(e, t * 1, 1)), i;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [format]
   */
  set format(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    var t = F(e) ? 0 : j(e, _.__wbindgen_export_1, _.__wbindgen_export_2), n = O;
    _.rendersessionoptions_set_format(this.__wbg_ptr, t, n);
  }
}
Symbol.dispose && (He.prototype[Symbol.dispose] = He.prototype.free);
const je = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_typstrenderer_free(r >>> 0, 1));
class ce {
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(ce.prototype);
    return t.__wbg_ptr = e, je.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, je.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_typstrenderer_free(e, 0);
  }
  constructor() {
    const e = _.typstrenderer_new();
    return this.__wbg_ptr = e >>> 0, je.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * @param {CreateSessionOptions | null} [options]
   * @returns {RenderSession}
   */
  create_session(e) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const s = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr);
      let o = 0;
      if (!F(e)) {
        if (N(e, Pe), e.__wbg_ptr === 0)
          throw new Error("Attempt to use a moved value");
        o = e.__destroy_into_raw();
      }
      _.typstrenderer_create_session(s, this.__wbg_ptr, o);
      var t = h().getInt32(s + 0, !0), n = h().getInt32(s + 4, !0), i = h().getInt32(s + 8, !0);
      if (i)
        throw P(n);
      return W.__wrap(t);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} session
   */
  reset(e) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const i = _.__wbindgen_add_to_stack_pointer(-16);
      if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      _.typstrenderer_reset(i, this.__wbg_ptr, e.__wbg_ptr);
      var t = h().getInt32(i + 0, !0), n = h().getInt32(i + 4, !0);
      if (n)
        throw P(t);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} session
   * @param {string} action
   * @param {Uint8Array} data
   */
  manipulate_data(e, t, n) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      const a = j(t, _.__wbindgen_export_1, _.__wbindgen_export_2), d = O, g = Ae(n, _.__wbindgen_export_1), f = O;
      _.typstrenderer_manipulate_data(o, this.__wbg_ptr, e.__wbg_ptr, a, d, g, f);
      var i = h().getInt32(o + 0, !0), s = h().getInt32(o + 4, !0);
      if (s)
        throw P(i);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {Uint8Array} artifact_content
   * @param {string} decoder
   * @returns {RenderSession}
   */
  session_from_artifact(e, t) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr);
      const a = Ae(e, _.__wbindgen_export_1), d = O, g = j(t, _.__wbindgen_export_1, _.__wbindgen_export_2), f = O;
      _.typstrenderer_session_from_artifact(o, this.__wbg_ptr, a, d, g, f);
      var n = h().getInt32(o + 0, !0), i = h().getInt32(o + 4, !0), s = h().getInt32(o + 8, !0);
      if (s)
        throw P(i);
      return W.__wrap(n);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} _v
   */
  load_glyph_pack(e) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const i = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr), _.typstrenderer_load_glyph_pack(i, this.__wbg_ptr, v(e));
      var t = h().getInt32(i + 0, !0), n = h().getInt32(i + 4, !0);
      if (n)
        throw P(t);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} session
   * @param {number} rect_lo_x
   * @param {number} rect_lo_y
   * @param {number} rect_hi_x
   * @param {number} rect_hi_y
   * @returns {string}
   */
  render_svg_diff(e, t, n, i, s) {
    let o, a;
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const f = _.__wbindgen_add_to_stack_pointer(-16);
      if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      _.typstrenderer_render_svg_diff(f, this.__wbg_ptr, e.__wbg_ptr, t, n, i, s);
      var d = h().getInt32(f + 0, !0), g = h().getInt32(f + 4, !0);
      return o = d, a = g, R(d, g);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16), _.__wbindgen_export_3(o, a, 1);
    }
  }
  /**
   * @param {RenderSession} session
   * @param {number | null} [parts]
   * @returns {string}
   */
  svg_data(e, t) {
    let n, i;
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const m = _.__wbindgen_add_to_stack_pointer(-16);
      if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      F(t) || p(t), _.typstrenderer_svg_data(m, this.__wbg_ptr, e.__wbg_ptr, F(t) ? 4294967297 : t >>> 0);
      var s = h().getInt32(m + 0, !0), o = h().getInt32(m + 4, !0), a = h().getInt32(m + 8, !0), d = h().getInt32(m + 12, !0), g = s, f = o;
      if (d)
        throw g = 0, f = 0, P(a);
      return n = g, i = f, R(g, f);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16), _.__wbindgen_export_3(n, i, 1);
    }
  }
  /**
   * @param {RenderSession} session
   * @returns {Array<any> | undefined}
   */
  get_customs(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
      throw new Error("Attempt to use a moved value");
    const t = _.typstrenderer_get_customs(this.__wbg_ptr, e.__wbg_ptr);
    return P(t);
  }
  /**
   * @param {RenderSession} session
   * @param {HTMLElement} root
   * @returns {boolean}
   */
  render_svg(e, t) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      _.typstrenderer_render_svg(o, this.__wbg_ptr, e.__wbg_ptr, v(t));
      var n = h().getInt32(o + 0, !0), i = h().getInt32(o + 4, !0), s = h().getInt32(o + 8, !0);
      if (s)
        throw P(i);
      return n !== 0;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} _w
   * @returns {Promise<TypstWorker>}
   */
  create_worker(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const t = _.typstrenderer_create_worker(this.__wbg_ptr, v(e));
    return P(t);
  }
  /**
   * @returns {WorkerBridge}
   */
  create_worker_bridge() {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const i = this.__destroy_into_raw(), s = _.__wbindgen_add_to_stack_pointer(-16);
      p(i), _.typstrenderer_create_worker_bridge(s, i);
      var e = h().getInt32(s + 0, !0), t = h().getInt32(s + 4, !0), n = h().getInt32(s + 8, !0);
      if (n)
        throw P(t);
      return ue.__wrap(e);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {RenderSession} ses
   * @param {HTMLElement} elem
   * @returns {Promise<IncrDomDocClient>}
   */
  mount_dom(e, t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
      throw new Error("Attempt to use a moved value");
    const n = _.typstrenderer_mount_dom(this.__wbg_ptr, e.__wbg_ptr, v(t));
    return P(n);
  }
  /**
   * @param {RenderSession} ses
   * @param {any} canvas
   * @param {RenderPageImageOptions | null} [options]
   * @returns {Promise<any>}
   */
  render_page_to_canvas(e, t, n) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    if (p(this.__wbg_ptr), N(e, W), e.__wbg_ptr === 0)
      throw new Error("Attempt to use a moved value");
    let i = 0;
    if (!F(n)) {
      if (N(n, Z), n.__wbg_ptr === 0)
        throw new Error("Attempt to use a moved value");
      i = n.__destroy_into_raw();
    }
    const s = _.typstrenderer_render_page_to_canvas(this.__wbg_ptr, e.__wbg_ptr, v(t), i);
    return P(s);
  }
}
Symbol.dispose && (ce.prototype[Symbol.dispose] = ce.prototype.free);
const ht = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_typstrendererbuilder_free(r >>> 0, 1));
class Ve {
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ht.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_typstrendererbuilder_free(e, 0);
  }
  constructor() {
    try {
      const i = _.__wbindgen_add_to_stack_pointer(-16);
      _.typstrendererbuilder_new(i);
      var e = h().getInt32(i + 0, !0), t = h().getInt32(i + 4, !0), n = h().getInt32(i + 8, !0);
      if (n)
        throw P(t);
      return this.__wbg_ptr = e >>> 0, ht.register(this, this.__wbg_ptr, this), this;
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @returns {Promise<TypstRenderer>}
   */
  build() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    const e = this.__destroy_into_raw();
    p(e);
    const t = _.typstrendererbuilder_build(e);
    return P(t);
  }
  /**
   * @param {any} _pack
   * @returns {Promise<void>}
   */
  add_glyph_pack(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const t = _.typstrendererbuilder_add_glyph_pack(this.__wbg_ptr, v(e));
    return P(t);
  }
  /**
   * @param {Uint8Array} _font_buffer
   * @returns {Promise<void>}
   */
  add_raw_font(e) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const t = _.typstrendererbuilder_add_raw_font(this.__wbg_ptr, v(e));
    return P(t);
  }
  /**
   * @param {any} _font
   * @param {any} _blob
   * @returns {Promise<void>}
   */
  add_lazy_font(e, t) {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const n = _.typstrendererbuilder_add_lazy_font(this.__wbg_ptr, v(e), v(t));
    return P(n);
  }
}
Symbol.dispose && (Ve.prototype[Symbol.dispose] = Ve.prototype.free);
const Gr = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_typstworker_free(r >>> 0, 1));
class Je {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, Gr.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_typstworker_free(e, 0);
  }
  /**
   * @param {string} _action
   * @param {Uint8Array} _data
   * @returns {Promise<any>}
   */
  manipulate_data(e, t) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const o = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr);
      const a = j(e, _.__wbindgen_export_1, _.__wbindgen_export_2), d = O;
      _.typstworker_manipulate_data(o, this.__wbg_ptr, a, d, v(t));
      var n = h().getInt32(o + 0, !0), i = h().getInt32(o + 4, !0), s = h().getInt32(o + 8, !0);
      if (s)
        throw P(i);
      return P(n);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @returns {Promise<any>}
   */
  get_pages_info() {
    if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
    p(this.__wbg_ptr);
    const e = _.typstworker_get_pages_info(this.__wbg_ptr);
    return P(e);
  }
  /**
   * @param {Uint8Array} _actions
   * @param {HTMLCanvasElement[]} _canvas_list
   * @param {RenderPageImageOptions[]} _data
   * @returns {Promise<any>}
   */
  render_canvas(e, t, n) {
    try {
      if (this.__wbg_ptr == 0) throw new Error("Attempt to use a moved value");
      const a = _.__wbindgen_add_to_stack_pointer(-16);
      p(this.__wbg_ptr);
      const d = Ae(e, _.__wbindgen_export_1), g = O, f = ct(t, _.__wbindgen_export_1), m = O, C = ct(n, _.__wbindgen_export_1), E = O;
      _.typstworker_render_canvas(a, this.__wbg_ptr, d, g, f, m, C, E);
      var i = h().getInt32(a + 0, !0), s = h().getInt32(a + 4, !0), o = h().getInt32(a + 8, !0);
      if (o)
        throw P(s);
      return P(i);
    } finally {
      _.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (Je.prototype[Symbol.dispose] = Je.prototype.free);
const mt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => _.__wbg_workerbridge_free(r >>> 0, 1));
class ue {
  constructor() {
    throw new Error("cannot invoke `new` directly");
  }
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(ue.prototype);
    return t.__wbg_ptr = e, mt.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, mt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    _.__wbg_workerbridge_free(e, 0);
  }
}
Symbol.dispose && (ue.prototype[Symbol.dispose] = ue.prototype.free);
const Kr = /* @__PURE__ */ new Set(["basic", "cors", "default"]);
async function Xr(r, e) {
  if (typeof Response == "function" && r instanceof Response) {
    if (typeof WebAssembly.instantiateStreaming == "function")
      try {
        return await WebAssembly.instantiateStreaming(r, e);
      } catch (n) {
        if (r.ok && Kr.has(r.type) && r.headers.get("Content-Type") !== "application/wasm")
          console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve Wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", n);
        else
          throw n;
      }
    const t = await r.arrayBuffer();
    return await WebAssembly.instantiate(t, e);
  } else {
    const t = await WebAssembly.instantiate(r, e);
    return t instanceof WebAssembly.Instance ? { instance: t, module: r } : t;
  }
}
function Gt() {
  const r = {};
  return r.wbg = {}, r.wbg.__wbg_appendChild_87a6cc0aeb132c06 = function() {
    return I(function(e, t) {
      const n = u(e).appendChild(u(t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_call_13410aac570ffff7 = function() {
    return I(function(e, t) {
      const n = u(e).call(u(t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_call_641db1bb5db5a579 = function() {
    return I(function(e, t, n, i) {
      const s = u(e).call(u(t), u(n), u(i));
      return v(s);
    }, arguments);
  }, r.wbg.__wbg_call_a5400b25a865cfd8 = function() {
    return I(function(e, t, n) {
      const i = u(e).call(u(t), u(n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_call_f1fd202ba222e0ec = function() {
    return I(function(e, t, n, i, s) {
      const o = u(e).call(u(t), u(n), u(i), u(s));
      return v(o);
    }, arguments);
  }, r.wbg.__wbg_clearRect_448c93ecc652d129 = function() {
    return l(function(e, t, n, i, s) {
      u(e).clearRect(t, n, i, s);
    }, arguments);
  }, r.wbg.__wbg_clientWidth_8a498b7a82cae772 = function() {
    return l(function(e) {
      const t = u(e).clientWidth;
      return p(t), t;
    }, arguments);
  }, r.wbg.__wbg_clip_2009d02864d41d14 = function() {
    return l(function(e, t) {
      u(e).clip(u(t));
    }, arguments);
  }, r.wbg.__wbg_clip_be4c2f92cf5ce700 = function() {
    return l(function(e, t) {
      u(e).clip(u(t));
    }, arguments);
  }, r.wbg.__wbg_cloneNode_79d46b18d5619863 = function() {
    return I(function(e) {
      const t = u(e).cloneNode();
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_content_a26016a510c10d06 = function() {
    return l(function(e) {
      const t = u(e).content;
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_createElement_4909dfa2011f2abe = function() {
    return I(function(e, t, n) {
      const i = u(e).createElement(R(t, n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_createImageBitmap_d788b02f79d6bb1c = function() {
    return I(function(e, t) {
      const n = u(e).createImageBitmap(u(t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_createImageBitmap_f0017ef8064b2e1d = function() {
    return I(function(e, t) {
      const n = u(e).createImageBitmap(u(t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_createObjectURL_c80225986d2b928b = function() {
    return I(function(e, t) {
      const n = URL.createObjectURL(u(t)), i = j(n, _.__wbindgen_export_1, _.__wbindgen_export_2), s = O;
      h().setInt32(e + 4, s, !0), h().setInt32(e + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_document_7d29d139bd619045 = function() {
    return l(function(e) {
      const t = u(e).document;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_drawImage_0fcff80aa0873c4e = function() {
    return I(function(e, t, n, i) {
      u(e).drawImage(u(t), n, i);
    }, arguments);
  }, r.wbg.__wbg_drawImage_231d6a0afcfbfc3d = function() {
    return I(function(e, t, n, i, s, o) {
      u(e).drawImage(u(t), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_drawImage_285f406fd1771a6d = function() {
    return I(function(e, t, n, i) {
      u(e).drawImage(u(t), n, i);
    }, arguments);
  }, r.wbg.__wbg_drawImage_4902f5efaec99256 = function() {
    return I(function(e, t, n, i, s, o) {
      u(e).drawImage(u(t), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_drawImage_69a008f28f3150a9 = function() {
    return I(function(e, t, n, i, s, o) {
      u(e).drawImage(u(t), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_drawImage_b8b2c11c9e042476 = function() {
    return I(function(e, t, n, i) {
      u(e).drawImage(u(t), n, i);
    }, arguments);
  }, r.wbg.__wbg_drawImage_e6fefc71cc7e1583 = function() {
    return I(function(e, t, n, i, s, o) {
      u(e).drawImage(u(t), n, i, s, o);
    }, arguments);
  }, r.wbg.__wbg_error_7534b8e9a36f1ab4 = function() {
    return l(function(e, t) {
      let n, i;
      try {
        n = e, i = t, console.error(R(e, t));
      } finally {
        _.__wbindgen_export_3(n, i, 1);
      }
    }, arguments);
  }, r.wbg.__wbg_fillRect_81988a39149f1e60 = function() {
    return l(function(e, t, n, i, s) {
      u(e).fillRect(t, n, i, s);
    }, arguments);
  }, r.wbg.__wbg_fillRect_a160edfa11fce49b = function() {
    return l(function(e, t, n, i, s) {
      u(e).fillRect(t, n, i, s);
    }, arguments);
  }, r.wbg.__wbg_fill_062772ac70d2abe5 = function() {
    return l(function(e, t) {
      u(e).fill(u(t));
    }, arguments);
  }, r.wbg.__wbg_fill_3aba41a9072d4324 = function() {
    return l(function(e, t, n) {
      u(e).fill(u(t), ut[n]);
    }, arguments);
  }, r.wbg.__wbg_fill_8fa14f899c13b2b8 = function() {
    return l(function(e, t, n) {
      u(e).fill(u(t), ut[n]);
    }, arguments);
  }, r.wbg.__wbg_fill_a12dd0da87e4b294 = function() {
    return l(function(e, t) {
      u(e).fill(u(t));
    }, arguments);
  }, r.wbg.__wbg_firstElementChild_27076cbfeed86254 = function() {
    return l(function(e) {
      const t = u(e).firstElementChild;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_firstElementChild_d38fc714a20581dc = function() {
    return l(function(e) {
      const t = u(e).firstElementChild;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_getAttribute_8bfaf67e99ed2ee3 = function() {
    return l(function(e, t, n, i) {
      const s = u(t).getAttribute(R(n, i));
      var o = F(s) ? 0 : j(s, _.__wbindgen_export_1, _.__wbindgen_export_2), a = O;
      h().setInt32(e + 4, a, !0), h().setInt32(e + 0, o, !0);
    }, arguments);
  }, r.wbg.__wbg_getContext_15e158d04230a6f6 = function() {
    return I(function(e, t, n) {
      const i = u(e).getContext(R(t, n));
      return F(i) ? 0 : v(i);
    }, arguments);
  }, r.wbg.__wbg_getContext_8b08935510bf607c = function() {
    return I(function(e, t, n) {
      const i = u(e).getContext(R(t, n));
      return F(i) ? 0 : v(i);
    }, arguments);
  }, r.wbg.__wbg_get_458e874b43b18b25 = function() {
    return I(function(e, t) {
      const n = Reflect.get(u(e), u(t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_globalCompositeOperation_a9343749f60fc3c3 = function() {
    return I(function(e, t) {
      const n = u(t).globalCompositeOperation, i = j(n, _.__wbindgen_export_1, _.__wbindgen_export_2), s = O;
      h().setInt32(e + 4, s, !0), h().setInt32(e + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_globalCompositeOperation_fca573f41169da73 = function() {
    return I(function(e, t) {
      const n = u(t).globalCompositeOperation, i = j(n, _.__wbindgen_export_1, _.__wbindgen_export_2), s = O;
      h().setInt32(e + 4, s, !0), h().setInt32(e + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_height_dcede88ab14fa083 = function() {
    return l(function(e) {
      const t = u(e).height;
      return p(t), t;
    }, arguments);
  }, r.wbg.__wbg_incrdomdocclient_new = function() {
    return l(function(e) {
      const t = _e.__wrap(e);
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_instanceof_CanvasRenderingContext2d_8c616198ec03b12f = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof CanvasRenderingContext2D;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_Element_162e4334c7d6f450 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof Element;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlCanvasElement_299c60950dbb3428 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof HTMLCanvasElement;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlDivElement_d375b5c79180c917 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof HTMLDivElement;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlElement_d60c51c41eb8699a = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof HTMLElement;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_HtmlTemplateElement_68e5e8157bcbd653 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof HTMLTemplateElement;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_ImageBitmap_e82e6a629ec3d3e8 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof ImageBitmap;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_OffscreenCanvasRenderingContext2d_6234c9d9a0c6958a = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof OffscreenCanvasRenderingContext2D;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_OffscreenCanvas_da9915e5258f24fb = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof OffscreenCanvas;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_Promise_3ec9e849bf41bdb6 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof Promise;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_SvgGraphicsElement_50bb1acb75ae606a = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof SVGGraphicsElement;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_SvgsvgElement_b0e42b8ec77c0d1e = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof SVGSVGElement;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_Window_12d20d558ef92592 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof Window;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_instanceof_WorkerGlobalScope_85d487cc157fd065 = function() {
    return l(function(e) {
      let t;
      try {
        t = u(e) instanceof WorkerGlobalScope;
      } catch {
        t = !1;
      }
      const n = t;
      return M(n), n;
    }, arguments);
  }, r.wbg.__wbg_lastElementChild_7b35e6537e048254 = function() {
    return l(function(e) {
      const t = u(e).lastElementChild;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_length_6bb7e81f9d7713e4 = function() {
    return l(function(e) {
      const t = u(e).length;
      return p(t), t;
    }, arguments);
  }, r.wbg.__wbg_log_6c7b5f4f00b8ce3f = function() {
    return l(function(e) {
      console.log(u(e));
    }, arguments);
  }, r.wbg.__wbg_log_7917fde260a8fd39 = function() {
    return l(function(e, t) {
      console.log(u(e), u(t));
    }, arguments);
  }, r.wbg.__wbg_measureText_ce63d5548118a11b = function() {
    return I(function(e, t, n) {
      const i = u(e).measureText(R(t, n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_new_19c25a3f2fa63a02 = function() {
    return l(function() {
      const e = new Object();
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_new_1b0539a012a993e2 = function() {
    return I(function() {
      const e = new Image();
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_new_1f3a344cf3123716 = function() {
    return l(function() {
      const e = new Array();
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_new_2e3c58a15f39f5f9 = function() {
    return l(function(e, t) {
      try {
        var n = { a: e, b: t }, i = (o, a) => {
          const d = n.a;
          n.a = 0;
          try {
            return Jr(d, n.b, o, a);
          } finally {
            n.a = d;
          }
        };
        const s = new Promise(i);
        return v(s);
      } finally {
        n.a = n.b = 0;
      }
    }, arguments);
  }, r.wbg.__wbg_new_8a6f238a6ece86ea = function() {
    return l(function() {
      const e = new Error();
      return v(e);
    }, arguments);
  }, r.wbg.__wbg_new_da9dc54c5db29dfa = function() {
    return l(function(e, t) {
      const n = new Error(R(e, t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_new_dd6c7b0f98f05245 = function() {
    return I(function(e, t) {
      const n = new OffscreenCanvas(e >>> 0, t >>> 0);
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_newfromslice_074c56947bd43469 = function() {
    return l(function(e, t) {
      const n = new Uint8Array(_t(e, t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_newnoargs_254190557c45b4ec = function() {
    return l(function(e, t) {
      const n = new Function(R(e, t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_newwithlength_a167dcc7aaa3ba77 = function() {
    return l(function(e) {
      const t = new Uint8Array(e >>> 0);
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_newwithpathstring_0dbb2ef6187c1fee = function() {
    return I(function(e, t) {
      const n = new Path2D(R(e, t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_newwithu8arraysequenceandoptions_2df1a97d9f42efa4 = function() {
    return I(function(e, t) {
      const n = new Blob(u(e), u(t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_nextElementSibling_4b565fe9b7cdb849 = function() {
    return l(function(e) {
      const t = u(e).nextElementSibling;
      return F(t) ? 0 : v(t);
    }, arguments);
  }, r.wbg.__wbg_push_330b2eb93e4e1212 = function() {
    return l(function(e, t) {
      const n = u(e).push(u(t));
      return p(n), n;
    }, arguments);
  }, r.wbg.__wbg_putImageData_0324b5d1f2cb58d9 = function() {
    return I(function(e, t, n, i) {
      u(e).putImageData(u(t), n, i);
    }, arguments);
  }, r.wbg.__wbg_putImageData_3413ec0b4386cceb = function() {
    return I(function(e, t, n, i) {
      u(e).putImageData(u(t), n, i);
    }, arguments);
  }, r.wbg.__wbg_queueMicrotask_25d0739ac89e8c88 = function() {
    return l(function(e) {
      queueMicrotask(u(e));
    }, arguments);
  }, r.wbg.__wbg_queueMicrotask_4488407636f5bf24 = function() {
    return l(function(e) {
      const t = u(e).queueMicrotask;
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_removeProperty_8912427f4d0f6361 = function() {
    return I(function(e, t, n, i) {
      const s = u(t).removeProperty(R(n, i)), o = j(s, _.__wbindgen_export_1, _.__wbindgen_export_2), a = O;
      h().setInt32(e + 4, a, !0), h().setInt32(e + 0, o, !0);
    }, arguments);
  }, r.wbg.__wbg_remove_fec7bce376b31b32 = function() {
    return l(function(e) {
      u(e).remove();
    }, arguments);
  }, r.wbg.__wbg_renderpageimageoptions_unwrap = function() {
    return l(function(e) {
      const t = Z.__unwrap(P(e));
      return p(t), t;
    }, arguments);
  }, r.wbg.__wbg_replaceWith_b871dd8b39227476 = function() {
    return I(function(e, t) {
      u(e).replaceWith(u(t));
    }, arguments);
  }, r.wbg.__wbg_resolve_4055c623acdd6a1b = function() {
    return l(function(e) {
      const t = Promise.resolve(u(e));
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_restore_1367a274a3a9c52f = function() {
    return l(function(e) {
      u(e).restore();
    }, arguments);
  }, r.wbg.__wbg_restore_45dead792d0756f6 = function() {
    return l(function(e) {
      u(e).restore();
    }, arguments);
  }, r.wbg.__wbg_revokeObjectURL_651b859c81bf1af0 = function() {
    return I(function(e, t) {
      URL.revokeObjectURL(R(e, t));
    }, arguments);
  }, r.wbg.__wbg_save_9c2af6c4f73acbb8 = function() {
    return l(function(e) {
      u(e).save();
    }, arguments);
  }, r.wbg.__wbg_save_e5fc517a51a9e73d = function() {
    return l(function(e) {
      u(e).save();
    }, arguments);
  }, r.wbg.__wbg_setAttribute_d1baf9023ad5696f = function() {
    return I(function(e, t, n, i, s) {
      u(e).setAttribute(R(t, n), R(i, s));
    }, arguments);
  }, r.wbg.__wbg_setLineDash_5c96338d7fd195f9 = function() {
    return I(function(e, t) {
      u(e).setLineDash(u(t));
    }, arguments);
  }, r.wbg.__wbg_setLineDash_7792494c18c240bc = function() {
    return I(function(e, t) {
      u(e).setLineDash(u(t));
    }, arguments);
  }, r.wbg.__wbg_setProperty_a4431938dd3e6945 = function() {
    return I(function(e, t, n, i, s) {
      u(e).setProperty(R(t, n), R(i, s));
    }, arguments);
  }, r.wbg.__wbg_setTransform_64b36b5b2fe7eb81 = function() {
    return I(function(e, t, n, i, s, o, a) {
      u(e).setTransform(t, n, i, s, o, a);
    }, arguments);
  }, r.wbg.__wbg_setTransform_e8b3bc7c8210c0e6 = function() {
    return I(function(e, t, n, i, s, o, a) {
      u(e).setTransform(t, n, i, s, o, a);
    }, arguments);
  }, r.wbg.__wbg_set_1353b2a5e96bc48c = function() {
    return l(function(e, t, n) {
      u(e).set(_t(t, n));
    }, arguments);
  }, r.wbg.__wbg_set_453345bcda80b89a = function() {
    return I(function(e, t, n) {
      const i = Reflect.set(u(e), u(t), u(n));
      return M(i), i;
    }, arguments);
  }, r.wbg.__wbg_set_90f6c0f7bd8c0415 = function() {
    return l(function(e, t, n) {
      u(e)[t >>> 0] = P(n);
    }, arguments);
  }, r.wbg.__wbg_setfillStyle_a9ad5b25cf62a5bc = function() {
    return l(function(e, t, n) {
      u(e).fillStyle = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setfillStyle_dae40abc0110b0d0 = function() {
    return l(function(e, t, n) {
      u(e).fillStyle = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setfont_73ce8e14831e02ab = function() {
    return l(function(e, t, n) {
      u(e).font = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setglobalCompositeOperation_e5ac8e2ee1998039 = function() {
    return I(function(e, t, n) {
      u(e).globalCompositeOperation = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setglobalCompositeOperation_e5ec3a5c4c8e9185 = function() {
    return I(function(e, t, n) {
      u(e).globalCompositeOperation = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setheight_4fce583024b2d088 = function() {
    return l(function(e, t) {
      u(e).height = t >>> 0;
    }, arguments);
  }, r.wbg.__wbg_setinnerHTML_34e240d6b8e8260c = function() {
    return l(function(e, t, n) {
      u(e).innerHTML = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setlineCap_8032972ab3563404 = function() {
    return l(function(e, t, n) {
      u(e).lineCap = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setlineCap_c4746ffc7c654931 = function() {
    return l(function(e, t, n) {
      u(e).lineCap = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setlineDashOffset_8e4721199e50ad71 = function() {
    return l(function(e, t) {
      u(e).lineDashOffset = t;
    }, arguments);
  }, r.wbg.__wbg_setlineDashOffset_f6bdd1bf33764013 = function() {
    return l(function(e, t) {
      u(e).lineDashOffset = t;
    }, arguments);
  }, r.wbg.__wbg_setlineJoin_2ff9aa15911a2360 = function() {
    return l(function(e, t, n) {
      u(e).lineJoin = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setlineJoin_30c59321fbc0bd3d = function() {
    return l(function(e, t, n) {
      u(e).lineJoin = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setlineWidth_069d571345379833 = function() {
    return l(function(e, t) {
      u(e).lineWidth = t;
    }, arguments);
  }, r.wbg.__wbg_setlineWidth_c640999c86cc71b6 = function() {
    return l(function(e, t) {
      u(e).lineWidth = t;
    }, arguments);
  }, r.wbg.__wbg_setmiterLimit_4b89ef6d3f97e1b0 = function() {
    return l(function(e, t) {
      u(e).miterLimit = t;
    }, arguments);
  }, r.wbg.__wbg_setmiterLimit_b4aa08d1762d6104 = function() {
    return l(function(e, t) {
      u(e).miterLimit = t;
    }, arguments);
  }, r.wbg.__wbg_setonerror_6eedc824b763863d = function() {
    return l(function(e, t) {
      u(e).onerror = u(t);
    }, arguments);
  }, r.wbg.__wbg_setonload_5f59eacd9a040d42 = function() {
    return l(function(e, t) {
      u(e).onload = u(t);
    }, arguments);
  }, r.wbg.__wbg_setsrc_953e439f2c70484a = function() {
    return l(function(e, t, n) {
      u(e).src = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setstrokeStyle_3c450999cfcdcd2f = function() {
    return l(function(e, t, n) {
      u(e).strokeStyle = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setstrokeStyle_e602cffe46d6113f = function() {
    return l(function(e, t, n) {
      u(e).strokeStyle = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_settype_298968e371b58a33 = function() {
    return l(function(e, t, n) {
      u(e).type = R(t, n);
    }, arguments);
  }, r.wbg.__wbg_setwidth_40a6ed203b92839d = function() {
    return l(function(e, t) {
      u(e).width = t >>> 0;
    }, arguments);
  }, r.wbg.__wbg_stack_0ed75d68575b0f3c = function() {
    return l(function(e, t) {
      const n = u(t).stack, i = j(n, _.__wbindgen_export_1, _.__wbindgen_export_2), s = O;
      h().setInt32(e + 4, s, !0), h().setInt32(e + 0, i, !0);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_GLOBAL_8921f820c2ce3f12 = function() {
    return l(function() {
      const e = typeof global > "u" ? null : global;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_GLOBAL_THIS_f0a4409105898184 = function() {
    return l(function() {
      const e = typeof globalThis > "u" ? null : globalThis;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_SELF_995b214ae681ff99 = function() {
    return l(function() {
      const e = typeof self > "u" ? null : self;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_static_accessor_WINDOW_cde3890479c675ea = function() {
    return l(function() {
      const e = typeof window > "u" ? null : window;
      return F(e) ? 0 : v(e);
    }, arguments);
  }, r.wbg.__wbg_stringify_b98c93d0a190446a = function() {
    return I(function(e) {
      const t = JSON.stringify(u(e));
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_stroke_1682cb55b83fabcd = function() {
    return l(function(e, t) {
      u(e).stroke(u(t));
    }, arguments);
  }, r.wbg.__wbg_stroke_84ffe7bcfdec57c3 = function() {
    return l(function(e, t) {
      u(e).stroke(u(t));
    }, arguments);
  }, r.wbg.__wbg_style_32a3c8393b46a115 = function() {
    return l(function(e) {
      const t = u(e).style;
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_then_b33a773d723afa3e = function() {
    return l(function(e, t, n) {
      const i = u(e).then(u(t), u(n));
      return v(i);
    }, arguments);
  }, r.wbg.__wbg_then_e22500defe16819f = function() {
    return l(function(e, t) {
      const n = u(e).then(u(t));
      return v(n);
    }, arguments);
  }, r.wbg.__wbg_transferToImageBitmap_ebacbcba8c548fc6 = function() {
    return I(function(e) {
      const t = u(e).transferToImageBitmap();
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_typstrenderer_new = function() {
    return l(function(e) {
      const t = ce.__wrap(e);
      return v(t);
    }, arguments);
  }, r.wbg.__wbg_warn_e2ada06313f92f09 = function() {
    return l(function(e) {
      console.warn(u(e));
    }, arguments);
  }, r.wbg.__wbg_wbindgencbdrop_eb10308566512b88 = function(e) {
    const t = u(e).original;
    if (t.cnt-- == 1)
      return t.a = 0, !0;
    const n = !1;
    return M(n), n;
  }, r.wbg.__wbg_wbindgendebugstring_99ef257a3ddda34d = function(e, t) {
    const n = Ne(u(t)), i = j(n, _.__wbindgen_export_1, _.__wbindgen_export_2), s = O;
    h().setInt32(e + 4, s, !0), h().setInt32(e + 0, i, !0);
  }, r.wbg.__wbg_wbindgenisfunction_8cee7dce3725ae74 = function(e) {
    const t = typeof u(e) == "function";
    return M(t), t;
  }, r.wbg.__wbg_wbindgenisundefined_c4b71d073b92f3c5 = function(e) {
    const t = u(e) === void 0;
    return M(t), t;
  }, r.wbg.__wbg_wbindgenjsvaleq_e6f2ad59ccae1b58 = function(e, t) {
    const n = u(e) === u(t);
    return M(n), n;
  }, r.wbg.__wbg_wbindgenstringget_0f16a6ddddef376f = function(e, t) {
    const n = u(t), i = typeof n == "string" ? n : void 0;
    var s = F(i) ? 0 : j(i, _.__wbindgen_export_1, _.__wbindgen_export_2), o = O;
    h().setInt32(e + 4, o, !0), h().setInt32(e + 0, s, !0);
  }, r.wbg.__wbg_wbindgenthrow_451ec1a8469d7eb6 = function(e, t) {
    throw new Error(R(e, t));
  }, r.wbg.__wbg_width_5671cc92dc930a91 = function() {
    return l(function(e) {
      return u(e).width;
    }, arguments);
  }, r.wbg.__wbg_width_cb2a94a03e71098c = function() {
    return l(function(e) {
      const t = u(e).width;
      return p(t), t;
    }, arguments);
  }, r.wbg.__wbindgen_cast_2241b6af4c4b2941 = function() {
    return l(function(e, t) {
      const n = R(e, t);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_cast_86466341b6bc6242 = function() {
    return l(function(e, t) {
      const n = at(e, t, 391, Vr);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_cast_b3a874c70d0fc95a = function() {
    return l(function(e, t) {
      const n = at(e, t, 391, Hr);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_cast_d6cd19b81560fd6e = function() {
    return l(function(e) {
      return v(e);
    }, arguments);
  }, r.wbg.__wbindgen_cast_f6a59d405c61762b = function() {
    return l(function(e, t) {
      const n = zr(e, t, 474, Nr);
      return v(n);
    }, arguments);
  }, r.wbg.__wbindgen_object_clone_ref = function(e) {
    const t = u(e);
    return v(t);
  }, r.wbg.__wbindgen_object_drop_ref = function(e) {
    P(e);
  }, r;
}
function Kt(r, e) {
  return _ = r.exports, Xt.__wbindgen_wasm_module = e, X = null, we = null, fe = null, _;
}
function Qr(r) {
  if (_ !== void 0) return _;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module: r } = r : console.warn("using deprecated parameters for `initSync()`; pass a single object instead"));
  const e = Gt();
  r instanceof WebAssembly.Module || (r = new WebAssembly.Module(r));
  const t = new WebAssembly.Instance(r, e);
  return Kt(t, r);
}
async function Xt(r) {
  if (_ !== void 0) return _;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module_or_path: r } = r : console.warn("using deprecated parameters for the initialization function; pass a single object instead")), typeof r > "u" && (r = Qt("typst_ts_renderer_bg.wasm", import.meta.url));
  const e = Gt();
  (typeof r == "string" || typeof Request == "function" && r instanceof Request || typeof URL == "function" && r instanceof URL) && (r = fetch(r));
  const { instance: t, module: n } = await Xr(await r, e);
  return Kt(t, n);
}
let Qt = async function(r, e) {
  throw new Error("Cannot import wasm module without importer: " + r + " " + e);
};
function Yt(r) {
  Qt = r;
}
let Yr = async function(r, e) {
  const t = new Function("m", "return import(m)"), { readFileSync: n } = await t("fs"), i = new URL(r, e);
  return await n(i).buffer;
};
const Zr = typeof process < "u" && process.versions != null && process.versions.node != null;
Zr && Yt(Yr);
const en = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  CreateSessionOptions: Pe,
  IncrDomDocClient: _e,
  PageInfo: Y,
  PagesInfo: ae,
  RenderPageImageOptions: Z,
  RenderSession: W,
  RenderSessionOptions: He,
  TypstRenderer: ce,
  TypstRendererBuilder: Ve,
  TypstWorker: Je,
  WorkerBridge: ue,
  default: Xt,
  initSync: Qr,
  renderer_build_info: Br,
  setImportWasmModule: Yt
}, Symbol.toStringTag, { value: "Module" }));
let c, pe = null;
function oe() {
  return (pe === null || pe.byteLength === 0) && (pe = new Uint8Array(c.memory.buffer)), pe;
}
let Ie = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 });
Ie.decode();
const tn = 2146435072;
let We = 0;
function rn(r, e) {
  return We += e, We >= tn && (Ie = new TextDecoder("utf-8", { ignoreBOM: !0, fatal: !0 }), Ie.decode(), We = e), Ie.decode(oe().subarray(r, r + e));
}
function q(r, e) {
  return r = r >>> 0, rn(r, e);
}
let J = new Array(128).fill(void 0);
J.push(void 0, null, !0, !1);
let me = J.length;
function k(r) {
  me === J.length && J.push(J.length + 1);
  const e = me;
  return me = J[e], J[e] = r, e;
}
function y(r) {
  return J[r];
}
function te(r, e) {
  try {
    return r.apply(this, e);
  } catch (t) {
    c.__wbindgen_export_0(k(t));
  }
}
function Ge(r, e) {
  return r = r >>> 0, oe().subarray(r / 1, r / 1 + e);
}
let be = null;
function nn() {
  return (be === null || be.byteLength === 0) && (be = new Uint32Array(c.memory.buffer)), be;
}
function sn(r, e) {
  return r = r >>> 0, nn().subarray(r / 4, r / 4 + e);
}
function on(r) {
  r < 132 || (J[r] = me, me = r);
}
function x(r) {
  const e = y(r);
  return on(r), e;
}
let A = 0;
const ye = new TextEncoder();
"encodeInto" in ye || (ye.encodeInto = function(r, e) {
  const t = ye.encode(r);
  return e.set(t), {
    read: r.length,
    written: t.length
  };
});
function L(r, e, t) {
  if (t === void 0) {
    const a = ye.encode(r), d = e(a.length, 1) >>> 0;
    return oe().subarray(d, d + a.length).set(a), A = a.length, d;
  }
  let n = r.length, i = e(n, 1) >>> 0;
  const s = oe();
  let o = 0;
  for (; o < n; o++) {
    const a = r.charCodeAt(o);
    if (a > 127) break;
    s[i + o] = a;
  }
  if (o !== n) {
    o !== 0 && (r = r.slice(o)), i = t(i, n, n = o + r.length * 3, 1) >>> 0;
    const a = oe().subarray(i + o, i + n), d = ye.encodeInto(r, a);
    o += d.written, i = t(i, n, o, 1) >>> 0;
  }
  return A = o, i;
}
let Q = null;
function w() {
  return (Q === null || Q.buffer.detached === !0 || Q.buffer.detached === void 0 && Q.buffer !== c.memory.buffer) && (Q = new DataView(c.memory.buffer)), Q;
}
function T(r) {
  return r == null;
}
function Ke(r) {
  const e = typeof r;
  if (e == "number" || e == "boolean" || r == null)
    return `${r}`;
  if (e == "string")
    return `"${r}"`;
  if (e == "symbol") {
    const i = r.description;
    return i == null ? "Symbol" : `Symbol(${i})`;
  }
  if (e == "function") {
    const i = r.name;
    return typeof i == "string" && i.length > 0 ? `Function(${i})` : "Function";
  }
  if (Array.isArray(r)) {
    const i = r.length;
    let s = "[";
    i > 0 && (s += Ke(r[0]));
    for (let o = 1; o < i; o++)
      s += ", " + Ke(r[o]);
    return s += "]", s;
  }
  const t = /\[object ([^\]]+)\]/.exec(toString.call(r));
  let n;
  if (t && t.length > 1)
    n = t[1];
  else
    return toString.call(r);
  if (n == "Object")
    try {
      return "Object(" + JSON.stringify(r) + ")";
    } catch {
      return "Object";
    }
  return r instanceof Error ? `${r.name}: ${r.message}
${r.stack}` : n;
}
const yt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry(
  (r) => {
    c.__wbindgen_export_4.get(r.dtor)(r.a, r.b);
  }
);
function _n(r, e, t, n) {
  const i = { a: r, b: e, cnt: 1, dtor: t }, s = (...o) => {
    i.cnt++;
    const a = i.a;
    i.a = 0;
    try {
      return n(a, i.b, ...o);
    } finally {
      --i.cnt === 0 ? (c.__wbindgen_export_4.get(i.dtor)(a, i.b), yt.unregister(i)) : i.a = a;
    }
  };
  return s.original = i, yt.register(s, i, i), s;
}
function an(r) {
  const e = c.get_font_info(k(r));
  return x(e);
}
function Xe(r, e) {
  if (!(r instanceof e))
    throw new Error(`expected instance of ${e.name}`);
}
function Zt(r, e) {
  const t = e(r.length * 1, 1) >>> 0;
  return oe().set(r, t / 1), A = r.length, t;
}
function cn(r, e) {
  r = r >>> 0;
  const t = w(), n = [];
  for (let i = r; i < r + 4 * e; i += 4)
    n.push(x(t.getUint32(i, !0)));
  return n;
}
function ke(r, e) {
  const t = e(r.length * 4, 4) >>> 0, n = w();
  for (let i = 0; i < r.length; i++)
    n.setUint32(t + 4 * i, k(r[i]), !0);
  return A = r.length, t;
}
function un(r, e, t) {
  c.__wbindgen_export_5(r, e, k(t));
}
function dn(r, e, t, n) {
  c.__wbindgen_export_6(r, e, k(t), k(n));
}
const vt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => c.__wbg_incrserver_free(r >>> 0, 1));
class K {
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(K.prototype);
    return t.__wbg_ptr = e, vt.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, vt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    c.__wbg_incrserver_free(e, 0);
  }
  /**
   * @param {boolean} attach
   */
  set_attach_debug_info(e) {
    c.incrserver_set_attach_debug_info(this.__wbg_ptr, e);
  }
  /**
   * @returns {Uint8Array | undefined}
   */
  current() {
    try {
      const n = c.__wbindgen_add_to_stack_pointer(-16);
      c.incrserver_current(n, this.__wbg_ptr);
      var e = w().getInt32(n + 0, !0), t = w().getInt32(n + 4, !0);
      let i;
      return e !== 0 && (i = Ge(e, t).slice(), c.__wbindgen_export_1(e, t * 1, 1)), i;
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  reset() {
    c.incrserver_reset(this.__wbg_ptr);
  }
}
Symbol.dispose && (K.prototype[Symbol.dispose] = K.prototype.free);
const ze = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => c.__wbg_proxycontext_free(r >>> 0, 1));
class de {
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(de.prototype);
    return t.__wbg_ptr = e, ze.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, ze.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    c.__wbg_proxycontext_free(e, 0);
  }
  /**
   * Creates a new `ProxyContext` instance.
   * @param {any} context
   */
  constructor(e) {
    const t = c.proxycontext_new(k(e));
    return this.__wbg_ptr = t >>> 0, ze.register(this, this.__wbg_ptr, this), this;
  }
  /**
   * Returns the JavaScript this.
   * @returns {any}
   */
  get context() {
    const e = c.proxycontext_context(this.__wbg_ptr);
    return x(e);
  }
  /**
   * A convenience function to untar a tarball and call a callback for each
   * entry.
   * @param {Uint8Array} data
   * @param {Function} cb
   */
  untar(e, t) {
    try {
      const s = c.__wbindgen_add_to_stack_pointer(-16), o = Zt(e, c.__wbindgen_export_2), a = A;
      c.proxycontext_untar(s, this.__wbg_ptr, o, a, k(t));
      var n = w().getInt32(s + 0, !0), i = w().getInt32(s + 4, !0);
      if (i)
        throw x(n);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (de.prototype[Symbol.dispose] = de.prototype.free);
const kt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => c.__wbg_typstcompileworld_free(r >>> 0, 1));
class ge {
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(ge.prototype);
    return t.__wbg_ptr = e, kt.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, kt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    c.__wbg_typstcompileworld_free(e, 0);
  }
  /**
   * @param {number} kind
   * @param {number} diagnostics_format
   * @returns {any}
   */
  compile(e, t) {
    try {
      const o = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompileworld_compile(o, this.__wbg_ptr, e, t);
      var n = w().getInt32(o + 0, !0), i = w().getInt32(o + 4, !0), s = w().getInt32(o + 8, !0);
      if (s)
        throw x(i);
      return x(n);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} kind
   * @returns {string | undefined}
   */
  title(e) {
    try {
      const o = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompileworld_title(o, this.__wbg_ptr, e);
      var t = w().getInt32(o + 0, !0), n = w().getInt32(o + 4, !0), i = w().getInt32(o + 8, !0), s = w().getInt32(o + 12, !0);
      if (s)
        throw x(i);
      let a;
      return t !== 0 && (a = q(t, n).slice(), c.__wbindgen_export_1(t, n * 1, 1)), a;
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} fmt
   * @param {number} diagnostics_format
   * @returns {any}
   */
  get_artifact(e, t) {
    try {
      const o = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompileworld_get_artifact(o, this.__wbg_ptr, e, t);
      var n = w().getInt32(o + 0, !0), i = w().getInt32(o + 4, !0), s = w().getInt32(o + 8, !0);
      if (s)
        throw x(i);
      return x(n);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {number} kind
   * @param {string} selector
   * @param {string | null} [field]
   * @returns {string}
   */
  query(e, t, n) {
    let i, s;
    try {
      const D = c.__wbindgen_add_to_stack_pointer(-16), $ = L(t, c.__wbindgen_export_2, c.__wbindgen_export_3), z = A;
      var o = T(n) ? 0 : L(n, c.__wbindgen_export_2, c.__wbindgen_export_3), a = A;
      c.typstcompileworld_query(D, this.__wbg_ptr, e, $, z, o, a);
      var d = w().getInt32(D + 0, !0), g = w().getInt32(D + 4, !0), f = w().getInt32(D + 8, !0), m = w().getInt32(D + 12, !0), C = d, E = g;
      if (m)
        throw C = 0, E = 0, x(f);
      return i = C, s = E, q(C, E);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16), c.__wbindgen_export_1(i, s, 1);
    }
  }
  /**
   * @param {IncrServer} state
   * @param {number} diagnostics_format
   * @returns {any}
   */
  incr_compile(e, t) {
    try {
      const o = c.__wbindgen_add_to_stack_pointer(-16);
      Xe(e, K), c.typstcompileworld_incr_compile(o, this.__wbg_ptr, e.__wbg_ptr, t);
      var n = w().getInt32(o + 0, !0), i = w().getInt32(o + 4, !0), s = w().getInt32(o + 8, !0);
      if (s)
        throw x(i);
      return x(n);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (ge.prototype[Symbol.dispose] = ge.prototype.free);
const xt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => c.__wbg_typstcompiler_free(r >>> 0, 1));
class le {
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(le.prototype);
    return t.__wbg_ptr = e, xt.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, xt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    c.__wbg_typstcompiler_free(e, 0);
  }
  reset() {
    try {
      const n = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompiler_reset(n, this.__wbg_ptr);
      var e = w().getInt32(n + 0, !0), t = w().getInt32(n + 4, !0);
      if (t)
        throw x(e);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {TypstFontResolver} fonts
   */
  set_fonts(e) {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      Xe(e, ee), c.typstcompiler_set_fonts(i, this.__wbg_ptr, e.__wbg_ptr);
      var t = w().getInt32(i + 0, !0), n = w().getInt32(i + 4, !0);
      if (n)
        throw x(t);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} inputs
   */
  set_inputs(e) {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompiler_set_inputs(i, this.__wbg_ptr, k(e));
      var t = w().getInt32(i + 0, !0), n = w().getInt32(i + 4, !0);
      if (n)
        throw x(t);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} path
   * @param {string} content
   * @returns {boolean}
   */
  add_source(e, t) {
    const n = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), i = A, s = L(t, c.__wbindgen_export_2, c.__wbindgen_export_3), o = A;
    return c.typstcompiler_add_source(this.__wbg_ptr, n, i, s, o) !== 0;
  }
  /**
   * @param {string} path
   * @param {Uint8Array} content
   * @returns {boolean}
   */
  map_shadow(e, t) {
    const n = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), i = A, s = Zt(t, c.__wbindgen_export_2), o = A;
    return c.typstcompiler_map_shadow(this.__wbg_ptr, n, i, s, o) !== 0;
  }
  /**
   * @param {string} path
   * @returns {boolean}
   */
  unmap_shadow(e) {
    const t = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), n = A;
    return c.typstcompiler_unmap_shadow(this.__wbg_ptr, t, n) !== 0;
  }
  reset_shadow() {
    c.typstcompiler_reset_shadow(this.__wbg_ptr);
  }
  /**
   * @returns {string[]}
   */
  get_loaded_fonts() {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompiler_get_loaded_fonts(i, this.__wbg_ptr);
      var e = w().getInt32(i + 0, !0), t = w().getInt32(i + 4, !0), n = cn(e, t).slice();
      return c.__wbindgen_export_1(e, t * 4, 4), n;
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} main_file_path
   * @returns {string}
   */
  get_ast(e) {
    let t, n;
    try {
      const f = c.__wbindgen_add_to_stack_pointer(-16), m = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), C = A;
      c.typstcompiler_get_ast(f, this.__wbg_ptr, m, C);
      var i = w().getInt32(f + 0, !0), s = w().getInt32(f + 4, !0), o = w().getInt32(f + 8, !0), a = w().getInt32(f + 12, !0), d = i, g = s;
      if (a)
        throw d = 0, g = 0, x(o);
      return t = d, n = g, q(d, g);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16), c.__wbindgen_export_1(t, n, 1);
    }
  }
  /**
   * @returns {any}
   */
  get_semantic_token_legend() {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompiler_get_semantic_token_legend(i, this.__wbg_ptr);
      var e = w().getInt32(i + 0, !0), t = w().getInt32(i + 4, !0), n = w().getInt32(i + 8, !0);
      if (n)
        throw x(t);
      return x(e);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} offset_encoding
   * @param {string | null} [file_path]
   * @param {string | null} [result_id]
   * @returns {object}
   */
  get_semantic_tokens(e, t, n) {
    try {
      const m = c.__wbindgen_add_to_stack_pointer(-16), C = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), E = A;
      var i = T(t) ? 0 : L(t, c.__wbindgen_export_2, c.__wbindgen_export_3), s = A, o = T(n) ? 0 : L(n, c.__wbindgen_export_2, c.__wbindgen_export_3), a = A;
      c.typstcompiler_get_semantic_tokens(m, this.__wbg_ptr, C, E, i, s, o, a);
      var d = w().getInt32(m + 0, !0), g = w().getInt32(m + 4, !0), f = w().getInt32(m + 8, !0);
      if (f)
        throw x(g);
      return x(d);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null} [root]
   * @param {string | null} [main_file_path]
   * @param {(Array<any>)[] | null} [inputs]
   * @returns {TypstCompileWorld}
   */
  snapshot(e, t, n) {
    try {
      const E = c.__wbindgen_add_to_stack_pointer(-16);
      var i = T(e) ? 0 : L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), s = A, o = T(t) ? 0 : L(t, c.__wbindgen_export_2, c.__wbindgen_export_3), a = A, d = T(n) ? 0 : ke(n, c.__wbindgen_export_2), g = A;
      c.typstcompiler_snapshot(E, this.__wbg_ptr, i, s, o, a, d, g);
      var f = w().getInt32(E + 0, !0), m = w().getInt32(E + 4, !0), C = w().getInt32(E + 8, !0);
      if (C)
        throw x(m);
      return ge.__wrap(f);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} fmt
   * @param {number} diagnostics_format
   * @returns {any}
   */
  get_artifact(e, t) {
    try {
      const o = c.__wbindgen_add_to_stack_pointer(-16), a = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), d = A;
      c.typstcompiler_get_artifact(o, this.__wbg_ptr, a, d, t);
      var n = w().getInt32(o + 0, !0), i = w().getInt32(o + 4, !0), s = w().getInt32(o + 8, !0);
      if (s)
        throw x(i);
      return x(n);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string | null | undefined} main_file_path
   * @param {(Array<any>)[] | null | undefined} inputs
   * @param {string} fmt
   * @param {number} diagnostics_format
   * @returns {any}
   */
  compile(e, t, n, i) {
    try {
      const C = c.__wbindgen_add_to_stack_pointer(-16);
      var s = T(e) ? 0 : L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), o = A, a = T(t) ? 0 : ke(t, c.__wbindgen_export_2), d = A;
      const E = L(n, c.__wbindgen_export_2, c.__wbindgen_export_3), D = A;
      c.typstcompiler_compile(C, this.__wbg_ptr, s, o, a, d, E, D, i);
      var g = w().getInt32(C + 0, !0), f = w().getInt32(C + 4, !0), m = w().getInt32(C + 8, !0);
      if (m)
        throw x(f);
      return x(g);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} main_file_path
   * @param {(Array<any>)[] | null | undefined} inputs
   * @param {string} selector
   * @param {string | null} [field]
   * @returns {string}
   */
  query(e, t, n, i) {
    let s, o;
    try {
      const B = c.__wbindgen_add_to_stack_pointer(-16), or = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), _r = A;
      var a = T(t) ? 0 : ke(t, c.__wbindgen_export_2), d = A;
      const ar = L(n, c.__wbindgen_export_2, c.__wbindgen_export_3), cr = A;
      var g = T(i) ? 0 : L(i, c.__wbindgen_export_2, c.__wbindgen_export_3), f = A;
      c.typstcompiler_query(B, this.__wbg_ptr, or, _r, a, d, ar, cr, g, f);
      var m = w().getInt32(B + 0, !0), C = w().getInt32(B + 4, !0), E = w().getInt32(B + 8, !0), D = w().getInt32(B + 12, !0), $ = m, z = C;
      if (D)
        throw $ = 0, z = 0, x(E);
      return s = $, o = z, q($, z);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16), c.__wbindgen_export_1(s, o, 1);
    }
  }
  /**
   * @returns {IncrServer}
   */
  create_incr_server() {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompiler_create_incr_server(i, this.__wbg_ptr);
      var e = w().getInt32(i + 0, !0), t = w().getInt32(i + 4, !0), n = w().getInt32(i + 8, !0);
      if (n)
        throw x(t);
      return K.__wrap(e);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {string} main_file_path
   * @param {(Array<any>)[] | null | undefined} inputs
   * @param {IncrServer} state
   * @param {number} diagnostics_format
   * @returns {any}
   */
  incr_compile(e, t, n, i) {
    try {
      const f = c.__wbindgen_add_to_stack_pointer(-16), m = L(e, c.__wbindgen_export_2, c.__wbindgen_export_3), C = A;
      var s = T(t) ? 0 : ke(t, c.__wbindgen_export_2), o = A;
      Xe(n, K), c.typstcompiler_incr_compile(f, this.__wbg_ptr, m, C, s, o, n.__wbg_ptr, i);
      var a = w().getInt32(f + 0, !0), d = w().getInt32(f + 4, !0), g = w().getInt32(f + 8, !0);
      if (g)
        throw x(d);
      return x(a);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
}
Symbol.dispose && (le.prototype[Symbol.dispose] = le.prototype.free);
const St = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => c.__wbg_typstcompilerbuilder_free(r >>> 0, 1));
class Qe {
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, St.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    c.__wbg_typstcompilerbuilder_free(e, 0);
  }
  constructor() {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompilerbuilder_new(i);
      var e = w().getInt32(i + 0, !0), t = w().getInt32(i + 4, !0), n = w().getInt32(i + 8, !0);
      if (n)
        throw x(t);
      return this.__wbg_ptr = e >>> 0, St.register(this, this.__wbg_ptr, this), this;
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  set_dummy_access_model() {
    try {
      const n = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstcompilerbuilder_set_dummy_access_model(n, this.__wbg_ptr);
      var e = w().getInt32(n + 0, !0), t = w().getInt32(n + 4, !0);
      if (t)
        throw x(e);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {any} context
   * @param {Function} mtime_fn
   * @param {Function} is_file_fn
   * @param {Function} real_path_fn
   * @param {Function} read_all_fn
   * @returns {Promise<void>}
   */
  set_access_model(e, t, n, i, s) {
    const o = c.typstcompilerbuilder_set_access_model(this.__wbg_ptr, k(e), k(t), k(n), k(i), k(s));
    return x(o);
  }
  /**
   * @param {any} context
   * @param {Function} real_resolve_fn
   * @returns {Promise<void>}
   */
  set_package_registry(e, t) {
    const n = c.typstcompilerbuilder_set_package_registry(this.__wbg_ptr, k(e), k(t));
    return x(n);
  }
  /**
   * @param {Uint8Array} data
   * @returns {Promise<void>}
   */
  add_raw_font(e) {
    const t = c.typstcompilerbuilder_add_raw_font(this.__wbg_ptr, k(e));
    return x(t);
  }
  /**
   * @param {any} font
   * @param {Function} blob
   * @returns {Promise<void>}
   */
  add_lazy_font(e, t) {
    const n = c.typstcompilerbuilder_add_lazy_font(this.__wbg_ptr, k(e), k(t));
    return x(n);
  }
  /**
   * @returns {Promise<TypstCompiler>}
   */
  build() {
    const e = this.__destroy_into_raw(), t = c.typstcompilerbuilder_build(e);
    return x(t);
  }
}
Symbol.dispose && (Qe.prototype[Symbol.dispose] = Qe.prototype.free);
const It = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => c.__wbg_typstfontresolver_free(r >>> 0, 1));
class ee {
  static __wrap(e) {
    e = e >>> 0;
    const t = Object.create(ee.prototype);
    return t.__wbg_ptr = e, It.register(t, t.__wbg_ptr, t), t;
  }
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, It.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    c.__wbg_typstfontresolver_free(e, 0);
  }
}
Symbol.dispose && (ee.prototype[Symbol.dispose] = ee.prototype.free);
const Rt = typeof FinalizationRegistry > "u" ? { register: () => {
}, unregister: () => {
} } : new FinalizationRegistry((r) => c.__wbg_typstfontresolverbuilder_free(r >>> 0, 1));
class Ye {
  __destroy_into_raw() {
    const e = this.__wbg_ptr;
    return this.__wbg_ptr = 0, Rt.unregister(this), e;
  }
  free() {
    const e = this.__destroy_into_raw();
    c.__wbg_typstfontresolverbuilder_free(e, 0);
  }
  constructor() {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstfontresolverbuilder_new(i);
      var e = w().getInt32(i + 0, !0), t = w().getInt32(i + 4, !0), n = w().getInt32(i + 8, !0);
      if (n)
        throw x(t);
      return this.__wbg_ptr = e >>> 0, Rt.register(this, this.__wbg_ptr, this), this;
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @param {Uint8Array} buffer
   * @returns {any}
   */
  get_font_info(e) {
    try {
      const s = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstfontresolverbuilder_get_font_info(s, this.__wbg_ptr, k(e));
      var t = w().getInt32(s + 0, !0), n = w().getInt32(s + 4, !0), i = w().getInt32(s + 8, !0);
      if (i)
        throw x(n);
      return x(t);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * Adds font data to the searcher.
   * @param {Uint8Array} buffer
   */
  add_raw_font(e) {
    try {
      const i = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstfontresolverbuilder_add_raw_font(i, this.__wbg_ptr, k(e));
      var t = w().getInt32(i + 0, !0), n = w().getInt32(i + 4, !0);
      if (n)
        throw x(t);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * Adds callback that loads font data lazily to the searcher.
   * `get_font_info` can be used to get the font info.
   * @param {any} font
   * @param {Function} blob
   */
  add_lazy_font(e, t) {
    try {
      const s = c.__wbindgen_add_to_stack_pointer(-16);
      c.typstfontresolverbuilder_add_lazy_font(s, this.__wbg_ptr, k(e), k(t));
      var n = w().getInt32(s + 0, !0), i = w().getInt32(s + 4, !0);
      if (i)
        throw x(n);
    } finally {
      c.__wbindgen_add_to_stack_pointer(16);
    }
  }
  /**
   * @returns {Promise<TypstFontResolver>}
   */
  build() {
    const e = this.__destroy_into_raw(), t = c.typstfontresolverbuilder_build(e);
    return x(t);
  }
}
Symbol.dispose && (Ye.prototype[Symbol.dispose] = Ye.prototype.free);
const gn = /* @__PURE__ */ new Set(["basic", "cors", "default"]);
async function ln(r, e) {
  if (typeof Response == "function" && r instanceof Response) {
    if (typeof WebAssembly.instantiateStreaming == "function")
      try {
        return await WebAssembly.instantiateStreaming(r, e);
      } catch (n) {
        if (r.ok && gn.has(r.type) && r.headers.get("Content-Type") !== "application/wasm")
          console.warn("`WebAssembly.instantiateStreaming` failed because your server does not serve Wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:\n", n);
        else
          throw n;
      }
    const t = await r.arrayBuffer();
    return await WebAssembly.instantiate(t, e);
  } else {
    const t = await WebAssembly.instantiate(r, e);
    return t instanceof WebAssembly.Instance ? { instance: t, module: r } : t;
  }
}
function er() {
  const r = {};
  return r.wbg = {}, r.wbg.__wbg_Error_e17e777aac105295 = function(e, t) {
    const n = Error(q(e, t));
    return k(n);
  }, r.wbg.__wbg_Number_998bea33bd87c3e0 = function(e) {
    return Number(y(e));
  }, r.wbg.__wbg_call_13410aac570ffff7 = function() {
    return te(function(e, t) {
      const n = y(e).call(y(t));
      return k(n);
    }, arguments);
  }, r.wbg.__wbg_call_a5400b25a865cfd8 = function() {
    return te(function(e, t, n) {
      const i = y(e).call(y(t), y(n));
      return k(i);
    }, arguments);
  }, r.wbg.__wbg_call_f1fd202ba222e0ec = function() {
    return te(function(e, t, n, i, s) {
      const o = y(e).call(y(t), y(n), y(i), y(s));
      return k(o);
    }, arguments);
  }, r.wbg.__wbg_done_75ed0ee6dd243d9d = function(e) {
    return y(e).done;
  }, r.wbg.__wbg_entries_2be2f15bd5554996 = function(e) {
    const t = Object.entries(y(e));
    return k(t);
  }, r.wbg.__wbg_error_7534b8e9a36f1ab4 = function(e, t) {
    let n, i;
    try {
      n = e, i = t, console.error(q(e, t));
    } finally {
      c.__wbindgen_export_1(n, i, 1);
    }
  }, r.wbg.__wbg_error_9eeaeffe2964d2ba = function(e, t, n) {
    console.error(y(e), y(t), y(n));
  }, r.wbg.__wbg_getTimezoneOffset_1e3ddc1382e7c8b0 = function(e) {
    return y(e).getTimezoneOffset();
  }, r.wbg.__wbg_get_0da715ceaecea5c8 = function(e, t) {
    const n = y(e)[t >>> 0];
    return k(n);
  }, r.wbg.__wbg_get_458e874b43b18b25 = function() {
    return te(function(e, t) {
      const n = Reflect.get(y(e), y(t));
      return k(n);
    }, arguments);
  }, r.wbg.__wbg_getwithrefkey_1dc361bd10053bfe = function(e, t) {
    const n = y(e)[y(t)];
    return k(n);
  }, r.wbg.__wbg_info_6cf68c1a86a92f6a = function(e) {
    console.info(y(e));
  }, r.wbg.__wbg_instanceof_ArrayBuffer_67f3012529f6a2dd = function(e) {
    let t;
    try {
      t = y(e) instanceof ArrayBuffer;
    } catch {
      t = !1;
    }
    return t;
  }, r.wbg.__wbg_instanceof_Map_ebb01a5b6b5ffd0b = function(e) {
    let t;
    try {
      t = y(e) instanceof Map;
    } catch {
      t = !1;
    }
    return t;
  }, r.wbg.__wbg_instanceof_Uint8Array_9a8378d955933db7 = function(e) {
    let t;
    try {
      t = y(e) instanceof Uint8Array;
    } catch {
      t = !1;
    }
    return t;
  }, r.wbg.__wbg_isArray_030cce220591fb41 = function(e) {
    return Array.isArray(y(e));
  }, r.wbg.__wbg_isSafeInteger_1c0d1af5542e102a = function(e) {
    return Number.isSafeInteger(y(e));
  }, r.wbg.__wbg_iterator_f370b34483c71a1c = function() {
    return k(Symbol.iterator);
  }, r.wbg.__wbg_length_186546c51cd61acd = function(e) {
    return y(e).length;
  }, r.wbg.__wbg_length_6bb7e81f9d7713e4 = function(e) {
    return y(e).length;
  }, r.wbg.__wbg_new_19c25a3f2fa63a02 = function() {
    const e = new Object();
    return k(e);
  }, r.wbg.__wbg_new_1f3a344cf3123716 = function() {
    const e = new Array();
    return k(e);
  }, r.wbg.__wbg_new_2e3c58a15f39f5f9 = function(e, t) {
    try {
      var n = { a: e, b: t }, i = (o, a) => {
        const d = n.a;
        n.a = 0;
        try {
          return dn(d, n.b, o, a);
        } finally {
          n.a = d;
        }
      };
      const s = new Promise(i);
      return k(s);
    } finally {
      n.a = n.b = 0;
    }
  }, r.wbg.__wbg_new_5a2ae4557f92b50e = function(e) {
    const t = new Date(y(e));
    return k(t);
  }, r.wbg.__wbg_new_638ebfaedbf32a5e = function(e) {
    const t = new Uint8Array(y(e));
    return k(t);
  }, r.wbg.__wbg_new_8a6f238a6ece86ea = function() {
    const e = new Error();
    return k(e);
  }, r.wbg.__wbg_new_da9dc54c5db29dfa = function(e, t) {
    const n = new Error(q(e, t));
    return k(n);
  }, r.wbg.__wbg_newfromslice_074c56947bd43469 = function(e, t) {
    const n = new Uint8Array(Ge(e, t));
    return k(n);
  }, r.wbg.__wbg_newfromslice_bd5b07f10e7e3c36 = function(e, t) {
    const n = new Uint32Array(sn(e, t));
    return k(n);
  }, r.wbg.__wbg_newnoargs_254190557c45b4ec = function(e, t) {
    const n = new Function(q(e, t));
    return k(n);
  }, r.wbg.__wbg_newwithargs_b8065bb443501079 = function(e, t, n, i) {
    const s = new Function(q(e, t), q(n, i));
    return k(s);
  }, r.wbg.__wbg_next_5b3530e612fde77d = function(e) {
    const t = y(e).next;
    return k(t);
  }, r.wbg.__wbg_next_692e82279131b03c = function() {
    return te(function(e) {
      const t = y(e).next();
      return k(t);
    }, arguments);
  }, r.wbg.__wbg_now_1e80617bcee43265 = function() {
    return Date.now();
  }, r.wbg.__wbg_prototypesetcall_3d4a26c1ed734349 = function(e, t, n) {
    Uint8Array.prototype.set.call(Ge(e, t), y(n));
  }, r.wbg.__wbg_proxycontext_new = function(e) {
    const t = de.__wrap(e);
    return k(t);
  }, r.wbg.__wbg_push_330b2eb93e4e1212 = function(e, t) {
    return y(e).push(y(t));
  }, r.wbg.__wbg_queueMicrotask_25d0739ac89e8c88 = function(e) {
    queueMicrotask(y(e));
  }, r.wbg.__wbg_queueMicrotask_4488407636f5bf24 = function(e) {
    const t = y(e).queueMicrotask;
    return k(t);
  }, r.wbg.__wbg_resolve_4055c623acdd6a1b = function(e) {
    const t = Promise.resolve(y(e));
    return k(t);
  }, r.wbg.__wbg_set_3f1d0b984ed272ed = function(e, t, n) {
    y(e)[x(t)] = x(n);
  }, r.wbg.__wbg_set_453345bcda80b89a = function() {
    return te(function(e, t, n) {
      return Reflect.set(y(e), y(t), y(n));
    }, arguments);
  }, r.wbg.__wbg_set_90f6c0f7bd8c0415 = function(e, t, n) {
    y(e)[t >>> 0] = x(n);
  }, r.wbg.__wbg_stack_0ed75d68575b0f3c = function(e, t) {
    const n = y(t).stack, i = L(n, c.__wbindgen_export_2, c.__wbindgen_export_3), s = A;
    w().setInt32(e + 4, s, !0), w().setInt32(e + 0, i, !0);
  }, r.wbg.__wbg_static_accessor_GLOBAL_8921f820c2ce3f12 = function() {
    const e = typeof global > "u" ? null : global;
    return T(e) ? 0 : k(e);
  }, r.wbg.__wbg_static_accessor_GLOBAL_THIS_f0a4409105898184 = function() {
    const e = typeof globalThis > "u" ? null : globalThis;
    return T(e) ? 0 : k(e);
  }, r.wbg.__wbg_static_accessor_SELF_995b214ae681ff99 = function() {
    const e = typeof self > "u" ? null : self;
    return T(e) ? 0 : k(e);
  }, r.wbg.__wbg_static_accessor_WINDOW_cde3890479c675ea = function() {
    const e = typeof window > "u" ? null : window;
    return T(e) ? 0 : k(e);
  }, r.wbg.__wbg_then_e22500defe16819f = function(e, t) {
    const n = y(e).then(y(t));
    return k(n);
  }, r.wbg.__wbg_typstcompiler_new = function(e) {
    const t = le.__wrap(e);
    return k(t);
  }, r.wbg.__wbg_typstfontresolver_new = function(e) {
    const t = ee.__wrap(e);
    return k(t);
  }, r.wbg.__wbg_value_dd9372230531eade = function(e) {
    const t = y(e).value;
    return k(t);
  }, r.wbg.__wbg_wbindgenbigintgetasi64_ac743ece6ab9bba1 = function(e, t) {
    const n = y(t), i = typeof n == "bigint" ? n : void 0;
    w().setBigInt64(e + 8, T(i) ? BigInt(0) : i, !0), w().setInt32(e + 0, !T(i), !0);
  }, r.wbg.__wbg_wbindgenbooleanget_3fe6f642c7d97746 = function(e) {
    const t = y(e), n = typeof t == "boolean" ? t : void 0;
    return T(n) ? 16777215 : n ? 1 : 0;
  }, r.wbg.__wbg_wbindgencbdrop_eb10308566512b88 = function(e) {
    const t = y(e).original;
    return t.cnt-- == 1 ? (t.a = 0, !0) : !1;
  }, r.wbg.__wbg_wbindgendebugstring_99ef257a3ddda34d = function(e, t) {
    const n = Ke(y(t)), i = L(n, c.__wbindgen_export_2, c.__wbindgen_export_3), s = A;
    w().setInt32(e + 4, s, !0), w().setInt32(e + 0, i, !0);
  }, r.wbg.__wbg_wbindgenin_d7a1ee10933d2d55 = function(e, t) {
    return y(e) in y(t);
  }, r.wbg.__wbg_wbindgenisbigint_ecb90cc08a5a9154 = function(e) {
    return typeof y(e) == "bigint";
  }, r.wbg.__wbg_wbindgenisfunction_8cee7dce3725ae74 = function(e) {
    return typeof y(e) == "function";
  }, r.wbg.__wbg_wbindgenisobject_307a53c6bd97fbf8 = function(e) {
    const t = y(e);
    return typeof t == "object" && t !== null;
  }, r.wbg.__wbg_wbindgenisstring_d4fa939789f003b0 = function(e) {
    return typeof y(e) == "string";
  }, r.wbg.__wbg_wbindgenisundefined_c4b71d073b92f3c5 = function(e) {
    return y(e) === void 0;
  }, r.wbg.__wbg_wbindgenjsvaleq_e6f2ad59ccae1b58 = function(e, t) {
    return y(e) === y(t);
  }, r.wbg.__wbg_wbindgenjsvallooseeq_9bec8c9be826bed1 = function(e, t) {
    return y(e) == y(t);
  }, r.wbg.__wbg_wbindgennumberget_f74b4c7525ac05cb = function(e, t) {
    const n = y(t), i = typeof n == "number" ? n : void 0;
    w().setFloat64(e + 8, T(i) ? 0 : i, !0), w().setInt32(e + 0, !T(i), !0);
  }, r.wbg.__wbg_wbindgenstringget_0f16a6ddddef376f = function(e, t) {
    const n = y(t), i = typeof n == "string" ? n : void 0;
    var s = T(i) ? 0 : L(i, c.__wbindgen_export_2, c.__wbindgen_export_3), o = A;
    w().setInt32(e + 4, o, !0), w().setInt32(e + 0, s, !0);
  }, r.wbg.__wbg_wbindgenthrow_451ec1a8469d7eb6 = function(e, t) {
    throw new Error(q(e, t));
  }, r.wbg.__wbindgen_cast_18dd738794d8ed40 = function(e, t) {
    const n = _n(e, t, 238, un);
    return k(n);
  }, r.wbg.__wbindgen_cast_2241b6af4c4b2941 = function(e, t) {
    const n = q(e, t);
    return k(n);
  }, r.wbg.__wbindgen_cast_4625c577ab2ec9ee = function(e) {
    const t = BigInt.asUintN(64, e);
    return k(t);
  }, r.wbg.__wbindgen_cast_9ae0607507abb057 = function(e) {
    return k(e);
  }, r.wbg.__wbindgen_cast_d6cd19b81560fd6e = function(e) {
    return k(e);
  }, r.wbg.__wbindgen_object_clone_ref = function(e) {
    const t = y(e);
    return k(t);
  }, r.wbg.__wbindgen_object_drop_ref = function(e) {
    x(e);
  }, r;
}
function tr(r, e) {
  return c = r.exports, rr.__wbindgen_wasm_module = e, Q = null, be = null, pe = null, c;
}
function fn(r) {
  if (c !== void 0) return c;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module: r } = r : console.warn("using deprecated parameters for `initSync()`; pass a single object instead"));
  const e = er();
  r instanceof WebAssembly.Module || (r = new WebAssembly.Module(r));
  const t = new WebAssembly.Instance(r, e);
  return tr(t, r);
}
async function rr(r) {
  if (c !== void 0) return c;
  typeof r < "u" && (Object.getPrototypeOf(r) === Object.prototype ? { module_or_path: r } = r : console.warn("using deprecated parameters for the initialization function; pass a single object instead")), typeof r > "u" && (r = nr("typst_ts_web_compiler_bg.wasm", import.meta.url));
  const e = er();
  (typeof r == "string" || typeof Request == "function" && r instanceof Request || typeof URL == "function" && r instanceof URL) && (r = fetch(r));
  const { instance: t, module: n } = await ln(await r, e);
  return tr(t, n);
}
let nr = async function(r, e) {
  throw new Error("Cannot import wasm module without importer: " + r + " " + e);
};
function ir(r) {
  nr = r;
}
let wn = async function(r, e) {
  const t = new Function("m", "return import(m)"), { readFileSync: n } = await t("fs"), i = new URL(r, e);
  return await n(i).buffer;
};
const pn = typeof process < "u" && process.versions != null && process.versions.node != null;
pn && ir(wn);
const sr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  IncrServer: K,
  ProxyContext: de,
  TypstCompileWorld: ge,
  TypstCompiler: le,
  TypstCompilerBuilder: Qe,
  TypstFontResolver: ee,
  TypstFontResolverBuilder: Ye,
  default: rr,
  get_font_info: an,
  initSync: fn,
  setImportWasmModule: ir
}, Symbol.toStringTag, { value: "Module" }));
let Et, Be;
function bn(r, e) {
  const t = Et || r();
  return Be !== void 0 ? Be : Be = (async () => (await t.init(e), Et = t))();
}
const hn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  createGlobalCompiler: bn
}, Symbol.toStringTag, { value: "Module" }));
export {
  Dr as TypstDocument
};
